---
title: "SecretSanta: flexible pipelines for automated prediction of secreted proteins."
author: "Anna Gogleva"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: Santa.bib
---  
Put somewhere: here we cover classical secretion mostly

**SecretSanta predict** is based on a family of wraper functions around [CBS tools](http://www.cbs.dtu.dk/services/). The wrappers are designed to work together. They produce output in a format recognosible by all of them. This alows to pipe results between individual predictors easily, as well as create flexible custom pipelines and compare outputs between several tools. 

To speed-up processing of large input fasta files (with several thousands of sequences) reccommended initial steps of the pipeline are automatically run as embarassingly parallel processes when number of input sequences exceedes a certain limit.

To avoid confusion, here and after names of CBS tools will be shown in bold magenta, corresponding R functions from the SecretSanta package would have standard code highlighting. For example:

+ <span style="color:magenta"> **signalp** </span> - CBS signalp;
+ ```signalp``` - SecretSanta wrap function around <span style="color:magenta"> **signalp** </span>;

Majority of the functions could be run in 2 modes: **starter** - in the case you initiate secretome pipeline with the above tool; and **piper** - for downstream/intermediate steps, so that the tool expects an output from another wraper function.

**SecretSanta annotate**
not implemented yet

**SecretSanta compare**
not implemented yet

**SecretSanta visualise**
not implemented yet

## Installation of external dependencies

For prediction of secreted proteins SecretSanta relies on a set of existing [CBS tools](http://www.cbs.dtu.dk/services/). Please install them and configure according to the listed requirements. Currently R intreface is implemented for the following tools:

\newline

### Tools for prediction of signal peptides and cleavage sites:
- **signalp-2.0** 
    + This version can run under IRIX, IRIX64, Linux, OSF1, SunOS.
    + Download stand alone signalp2.0 http://www.cbs.dtu.dk/cgi-bin/sw_request?signalp+2.0
    + Unpack the archive
    ```{sh eval = FALSE}
    tar -zxvf signalp-2.0.Linux.tar.Z
    cd signalp-2.0
    ```
    + Edit "General settings" at the top of the signalp file. Set value of 'SIGNALP' variable to be path to your signalp2.0 directory. Other variables usually do not require changes. We will not use plotting functions from signalp, so gnuplot, ppmtogif and ghostview are not requred. For more details please check ```signalp-2.0.readme```. 

\newline

- **signalp-3.0** 
    + This version will run on the most common UNIX platforms.
    + Download stand alone signalp3.0 http://www.cbs.dtu.dk/cgi-bin/sw_request?signalp+3.0
    + Unpack the archive
    ```{sh eval = FALSE}
    tar -zxvf signalp-3.0.Linux.tar.Z
    cd signalp-3.0
    ```
    + Similar to signalp2, edit "General settings" at the top of the signalp file. Set value of 'SIGNALP' variable to be path to your signalp3.0 directory. Other variables usually do not require changes. We will not use plotting functions from signalp, so gnuplot, ppmtogif and ghostview are not requred. For more details please check ```signalp-3.0.readme```.
    
\newline
 
- **signalp-4.1** - the most recent version
    + This version can run under Windows, OS X (Macintosh) and Linux.
    + Download stand alone signalp4.0 http://www.cbs.dtu.dk/cgi-bin/sw_request?signalp+4.0
    + Unpack the archive
    ```{sh eval = FALSE}
    tar -zxvf signalp-4.1.Linux.tar.Z
    cd signalp-4.1
    ```
    + Edit "General settings" at the top of the signalp file. Set values for 'SIGNALP' and 'outputDir' variables.
    For more details please check ```signalp-4.1.readme```.

\newline

### Tools for prediction of protein subcellular localisation:    
    
- **taretp-1.1**
    + tatgetp1.1 will run on the most common UNIX platforms  
    + Download stand alone targetp http://www.cbs.dtu.dk/cgi-bin/nph-sw_request?targetp
    + Unpack the archive:
    ```{sh eval = FALSE}
    tar -zxvf targetp-1.1b.Linux.tar.Z 
    cd targetp-1.1
    ```
    + Edit the paragraph labeled "GENERAL SETTINGS, customise" at the top of the targetp file.
    Set values for 'TARGETP' and 'TMP' variables. Ensure, that the path to targetp does not exceed 60 characters,
    otherwise targetp might fail.

\newline

- **WoLFPsort**
    + Clone WoLFPsort
    ```{sh eval = FALSE}
    git clone https://github.com/fmaguire/WoLFPSort.git
    cd WoLFPSort
    ```
    + Copy the binaries from the appropriate platform specific binary directory ```./bin/binByPlatform/binary-?``` to ```./bin/```
    + For more details please check the ```INSTALL``` file.
    
\newline

### Tools for prediction of transmembrane domains
- **tmhmm-2.0**
    + tmhmm-2.0 will run on the most common UNIX platforms  
    + Download stand alone tmhmm (http://www.cbs.dtu.dk/cgi-bin/nph-sw_request?tmhmm)
    + Upack the archive:
    ```{sh eval = FALSE}
    tar -zxvf tmhmm-2.0c.Linux.tar.gz
    cd tmhmm-2.0c
    ```
    + Set correct path for perl 5.x in the first line of ```bin/tmhmm``` and ```bin/tmhmmformat.pl``` scripts.
    + For more details please check the ```README``` file.
    

\newline

## Check that all the external dependencies are present and executable in principle:
```{r, message = FALSE}
library(SecretSanta)
```
To do this create a 2-column space-separated text file with listed paths for all the external dependencies. Sample input file listing paths to external dependencies:

```{r, echo = FALSE}
pp <-  suppressMessages(readr::read_delim(system.file("extdata", "santa_paths",
                                                      package="SecretSanta"),
                                          delim = ' ',
                                          col_names = FALSE))
names(pp) <- c('tool', 'path')

knitr::kable(pp)
```

For CBS tool names please use simple character strings without version numbers. The only exception here are different versions of <span style="color:magenta">**signalp**</span>, which require version number in one-digit format in the first column to be distinguishable.

Please note that aliases will not work when tools are called from R and changing ```$PATH``` variable is not necessary.

To check that all the required dependencies are installed correctly and can be executed in principle, run ```manage_paths``` function. Note, that ```manage_paths``` is case insensitive and will  convert all the tool names to the lower case. 

```{r, echo = TRUE, collapse = TRUE, tidy = TRUE}
santa_paths <- manage_paths(system.file("extdata", "sample_paths", package = "SecretSanta"))
```

## Signalp

<span style="color:magenta"> **Signalp** </span> is a softaware tool to predict classical signal peptides and cleavage sites [@Nielsen1997] in eukaryotes and bacteria. The method combines prediction of signal peptides and cleavage cites based on a combination of artificial neural networks as well as hidden Markov models. The latter allows to distinguish between signal peptides and non-cleaved signal anchors. 

Currently ```signalp``` function from SecretSanta provides an interface for the three recent versions of <span style="color:magenta"> **signalp** </span>. The reason behind this is that <span style="color:magenta"> **signalp4**  and **signalp4.1** </span> are not sensitive enough to predict some classes of secreted oomycete and fungal effectors [@Sperschneider2015]. Thus, providing acces to the older versions provide greater flexibility for building versatile prediction pipelines for a wider taxonomic range of species. 

- signalp2.0 [@Nielsen1998]
- signalp3.0 []
- signalp4.1

In case of large input files SecretSanta runs signalp as an embarassingly parallel process, suitable for clusters.

To run signalp prediction:

## Targetp

## TMHMM

## WolfPsort

## Check for terminal ER-retention motifs

## M-slicer

## Build pipelines, simple workflow

## Build pipelines, stringent workflow



Note the various macros within the `vignette` section of the metadata block above. These are required in order to instruct R how to build the vignette. Note that you should change the `title` field and the `\VignetteIndexEntry` to match the title of your vignette.

## Styles

The `html_vignette` template includes a basic CSS theme. To override this theme you can specify your own CSS in the document metadata as follows:

    output: 
      rmarkdown::html_vignette:
        css: mystyles.css

## Figures

The figure sizes have been customised so that you can easily put two images side-by-side. 

```{r, fig.show='hold'}
plot(1:10)
plot(10:1)
```

You can enable figure captions by `fig_caption: yes` in YAML:

    output:
      rmarkdown::html_vignette:
        fig_caption: yes

Then you can use the chunk option `fig.cap = "Your figure caption."` in **knitr**.

## More Examples

You can write math expressions, e.g. $Y = X\beta + \epsilon$, footnotes^[A footnote here.], and tables, e.g. using `knitr::kable()`.

```{r, echo=FALSE, results='asis'}
knitr::kable(head(mtcars, 10))
```

### References
