<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Anna Gogleva" />

<meta name="date" content="2017-08-21" />

<title>SecretSanta: flexible pipelines for automated prediction of secreted proteins.</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">SecretSanta: flexible pipelines for automated prediction of secreted proteins.</h1>
<h4 class="author"><em>Anna Gogleva</em></h4>
<h4 class="date"><em>2017-08-21</em></h4>



<p>Put somewhere: here we cover classical secretion mostly</p>
<p><strong>SecretSanta predict</strong> is based on a family of wraper functions around <a href="http://www.cbs.dtu.dk/services/">CBS tools</a>. The wrappers are designed to work together. They produce output in a format recognosible by all of them. This alows to pipe results between individual predictors easily, as well as create flexible custom pipelines and compare outputs between several tools.</p>
<p>To speed-up processing of large input fasta files (with several thousands of sequences) reccommended initial steps of the pipeline are automatically run as embarassingly parallel processes when number of input sequences exceedes a certain limit.</p>
<p>To avoid confusion, here and after names of external tools will be shown in bold magenta, corresponding R functions from the SecretSanta package would have standard code highlighting. For example:</p>
<ul>
<li><strong>signalp</strong> - CBS signalp and it’s associated derivative files;</li>
<li><code>signalp</code> - SecretSanta wrap function around <strong>signalp</strong>;</li>
</ul>
<p>Majority of the functions could be run in 2 modes: <em>starter</em> - in the case you initiate secretome pipeline with the above tool; and <em>piper</em> - for downstream/intermediate steps, so that the tool expects an output from another wraper function.</p>
<p><strong>SecretSanta annotate</strong> not implemented yet</p>
<p><strong>SecretSanta compare</strong> not implemented yet</p>
<p><strong>SecretSanta visualise</strong> not implemented yet</p>
<div id="installation-of-external-dependencies" class="section level2">
<h2>Installation of external dependencies</h2>
<p>For prediction of secreted proteins SecretSanta relies on a set of existing <a href="http://www.cbs.dtu.dk/services/">CBS tools</a>. Please install them and configure according to the listed requirements. Currently R intreface is implemented for the following tools:</p>
<p></p>
<div id="tools-for-prediction-of-signal-peptides-and-cleavage-sites" class="section level3">
<h3>Tools for prediction of signal peptides and cleavage sites:</h3>
<ul>
<li><strong>signalp-2.0</strong>
<ul>
<li>This version can run under IRIX, IRIX64, Linux, OSF1, SunOS.</li>
<li>Download stand alone signalp2.0 <a href="http://www.cbs.dtu.dk/cgi-bin/sw_request?signalp+2.0" class="uri">http://www.cbs.dtu.dk/cgi-bin/sw_request?signalp+2.0</a></li>
<li>Unpack the archive</li>
</ul>
<div class="sourceCode"><pre class="sourceCode sh"><code class="sourceCode bash"><span class="kw">tar</span> -zxvf signalp-2.0.Linux.tar.Z
<span class="kw">cd</span> signalp-2.0</code></pre></div>
<ul>
<li>Edit “General settings” at the top of the <strong>signalp</strong> file. Set value of ‘SIGNALP’ variable to be path to your <strong>signalp-2.0</strong> directory. Other variables usually do not require changes. We will not use plotting functions from signalp, so gnuplot, ppmtogif and ghostview are not requred. For more details please check <code>signalp-2.0.readme</code>.</li>
</ul></li>
</ul>
<p></p>
<ul>
<li><strong>signalp-3.0</strong>
<ul>
<li>This version will run on the most common UNIX platforms.</li>
<li>Download stand alone signalp3.0 <a href="http://www.cbs.dtu.dk/cgi-bin/sw_request?signalp+3.0" class="uri">http://www.cbs.dtu.dk/cgi-bin/sw_request?signalp+3.0</a></li>
<li>Unpack the archive</li>
</ul>
<div class="sourceCode"><pre class="sourceCode sh"><code class="sourceCode bash"><span class="kw">tar</span> -zxvf signalp-3.0.Linux.tar.Z
<span class="kw">cd</span> signalp-3.0</code></pre></div>
<ul>
<li>Similar to <strong>signalp-2.0</strong>, edit “General settings” at the top of the signalp file. Set value of ‘SIGNALP’ variable to be path to your <strong>signalp-3.0</strong> directory. Other variables usually do not require changes. We will not use plotting functions from <strong>signalp</strong>, so gnuplot, ppmtogif and ghostview are not requred. For more details please check <code>signalp-3.0.readme</code>.</li>
</ul></li>
</ul>
<p></p>
<ul>
<li><strong>signalp-4.1</strong> - the most recent version
<ul>
<li>This version can run under Windows, OS X (Macintosh) and Linux.</li>
<li>Download stand alone signalp4.0 <a href="http://www.cbs.dtu.dk/cgi-bin/sw_request?signalp+4.0" class="uri">http://www.cbs.dtu.dk/cgi-bin/sw_request?signalp+4.0</a></li>
<li>Unpack the archive</li>
</ul>
<div class="sourceCode"><pre class="sourceCode sh"><code class="sourceCode bash"><span class="kw">tar</span> -zxvf signalp-4.1.Linux.tar.Z
<span class="kw">cd</span> signalp-4.1</code></pre></div>
<ul>
<li>Edit “General settings” at the top of the <strong>signalp</strong> file. Set values for ‘SIGNALP’ and ‘outputDir’ variables. For more details please check <code>signalp-4.1.readme</code>.</li>
</ul></li>
</ul>
<p></p>
</div>
<div id="tools-for-prediction-of-protein-subcellular-localisation" class="section level3">
<h3>Tools for prediction of protein subcellular localisation:</h3>
<ul>
<li><strong>taretp-1.1</strong>
<ul>
<li><strong>tatgetp-1.1</strong> will run on the most common UNIX platforms<br />
</li>
<li>Download stand alone targetp <a href="http://www.cbs.dtu.dk/cgi-bin/nph-sw_request?targetp" class="uri">http://www.cbs.dtu.dk/cgi-bin/nph-sw_request?targetp</a></li>
<li>Unpack the archive:</li>
</ul>
<div class="sourceCode"><pre class="sourceCode sh"><code class="sourceCode bash"><span class="kw">tar</span> -zxvf targetp-1.1b.Linux.tar.Z 
<span class="kw">cd</span> targetp-1.1</code></pre></div>
<ul>
<li>Edit the paragraph labeled “GENERAL SETTINGS, customise” at the top of the <strong>targetp</strong> file. Set values for ‘TARGETP’ and ‘TMP’ variables. Ensure, that the path to targetp does not exceed 60 characters, otherwise <strong>targetp-1.1</strong> might fail.</li>
</ul></li>
</ul>
<p></p>
<ul>
<li><strong>WoLFPsort</strong>
<ul>
<li>Clone WoLFPsort</li>
</ul>
<div class="sourceCode"><pre class="sourceCode sh"><code class="sourceCode bash"><span class="kw">git</span> clone https://github.com/fmaguire/WoLFPSort.git
<span class="kw">cd</span> WoLFPSort</code></pre></div>
<ul>
<li>Copy the binaries from the appropriate platform specific binary directory <code>./bin/binByPlatform/binary-?</code> to <code>./bin/</code></li>
<li>For more details please check the <code>INSTALL</code> file.</li>
</ul></li>
</ul>
<p></p>
</div>
<div id="tools-for-prediction-of-transmembrane-domains" class="section level3">
<h3>Tools for prediction of transmembrane domains</h3>
<ul>
<li><strong>tmhmm-2.0</strong>
<ul>
<li><strong>tmhmm-2.0</strong> will run on the most common UNIX platforms<br />
</li>
<li>Download stand alone tmhmm (<a href="http://www.cbs.dtu.dk/cgi-bin/nph-sw_request?tmhmm" class="uri">http://www.cbs.dtu.dk/cgi-bin/nph-sw_request?tmhmm</a>)</li>
<li>Upack the archive:</li>
</ul>
<div class="sourceCode"><pre class="sourceCode sh"><code class="sourceCode bash"><span class="kw">tar</span> -zxvf tmhmm-2.0c.Linux.tar.gz
<span class="kw">cd</span> tmhmm-2.0c</code></pre></div>
<ul>
<li>Set correct path for perl 5.x in the first line of <code>bin/tmhmm</code> and <code>bin/tmhmmformat.pl</code> scripts.</li>
<li>For more details please check the <code>README</code> file.</li>
</ul></li>
</ul>
<p></p>
</div>
</div>
<div id="check-that-all-the-external-dependencies-are-present-and-executable-in-principle" class="section level2">
<h2>Check that all the external dependencies are present and executable in principle:</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(SecretSanta)</code></pre></div>
<p>To do this create a 2-column space-separated text file with listed paths for all the external dependencies. Sample input file listing paths to external dependencies:</p>
<table>
<thead>
<tr class="header">
<th align="left">tool</th>
<th align="left">path</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">signalp4</td>
<td align="left">/home/tools/signalp-4.1/signalp</td>
</tr>
<tr class="even">
<td align="left">signalp3</td>
<td align="left">/home/tools/signalp-3.0/signalp</td>
</tr>
<tr class="odd">
<td align="left">signalp2</td>
<td align="left">/home/tools/signalp-2.0/signalp</td>
</tr>
<tr class="even">
<td align="left">targetp</td>
<td align="left">/home/tools/targetp-1.1/targetp</td>
</tr>
<tr class="odd">
<td align="left">TMHMM</td>
<td align="left">/home/tools/tmhmm-2.0c/bin/tmhmm</td>
</tr>
<tr class="even">
<td align="left">WoLFPSORT</td>
<td align="left">/home/tools/WoLFPSort-master/bin/runWolfPsortSummary</td>
</tr>
</tbody>
</table>
<p>For CBS tool names please use simple character strings without version numbers. The only exception here are different versions of <strong>signalp</strong>, which require version number in one-digit format in the first column to be distinguishable.</p>
<p>Please note that aliases will not work when tools are called from R and changing <code>$PATH</code> variable is not necessary.</p>
<p>To check that all the required dependencies are installed correctly and can be executed in principle, run <code>manage_paths</code> function. Note, that <code>manage_paths</code> is case insensitive and will convert all the tool names to the lower case.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">santa_paths &lt;-<span class="st"> </span><span class="kw">manage_paths</span>(<span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>,
                                        <span class="st">&quot;sample_paths&quot;</span>,
                                        <span class="dt">package =</span> <span class="st">&quot;SecretSanta&quot;</span>))
## All paths are valid
## signalp2 test run completed
## signalp3 test run completed
## signalp4 test run completed
## targetp test run completed
## tmhmm test run completed
## wolfpsort test run completed</code></pre></div>
</div>
<div id="signalp" class="section level2">
<h2>Signalp</h2>
<p><strong>signalp</strong> is a softaware tool to predict classical signal peptides and cleavage sites <span class="citation">(Henrik Nielsen and Heijne 1997)</span> in eukaryotes and bacteria. The method combines prediction of signal peptides and cleavage cites based on a combination of artificial neural networks as well as hidden Markov models. The latter allows to distinguish between signal peptides and non-cleaved signal anchors.</p>
<p>Currently <code>signalp</code> function from SecretSanta provides an interface for the three recent versions of <strong>signalp</strong>. The reason behind this is that <strong>signalp-4</strong> and <strong>signalp-4.1</strong> are not sensitive enough to predict certain classes of secreted oomycete and fungal effectors <span class="citation">(Sperschneider J 2015)</span>. Thus, providing acces to the older versions provide greater flexibility for building versatile prediction pipelines for a wider taxonomic range of species.</p>
<ul>
<li><strong>signalp-2.0</strong> <span class="citation">(Nielsen and Krogh 1998)</span></li>
<li><strong>signalp-3.0</strong> <span class="citation">(Dyrl�v Bendtsen and Brunak 2004)</span></li>
<li><strong>signalp-4.0</strong> <span class="citation">(Thomas Nordahl Petersen and Nielsen. 2011)</span></li>
</ul>
<p>In case of large input files SecretSanta runs signalp as an embarassingly parallel process, suitable for clusters. To run signalp prediction, first read fasta file with amino acid sequences and store it’s contents in a separate variable:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">aa &lt;-<span class="st"> </span><span class="kw">readAAStringSet</span>(<span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>,
                                  <span class="st">&quot;sample_prot_100.fasta&quot;</span>,
                                  <span class="dt">package =</span> <span class="st">&quot;SecretSanta&quot;</span>))</code></pre></div>
<p>Initialise object of CBSResult class with <code>aa</code> as <code>in_fasta</code> attribute.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">inp &lt;-<span class="st"> </span><span class="kw">CBSResult</span>(<span class="dt">in_fasta =</span> aa)</code></pre></div>
<p>Since this is the first step in our secretome prediction pipeline, we will run <code>signalp</code> in a <em>starter</em> mode. Here we select version number 2 (<strong>signalp-2.0</strong>) and specify <code>santa_path</code> object generated earlier as a path container for external dependencies:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">step1_sp2 &lt;-<span class="st"> </span><span class="kw">signalp</span>(inp,
                     <span class="dt">version =</span> <span class="dv">2</span>,
                     <span class="dt">organism_type =</span> <span class="st">'euk'</span>, 
                     <span class="dt">run_mode =</span> <span class="st">&quot;starter&quot;</span>,
                     <span class="dt">paths =</span> santa_paths)
## Version used... signalp2
## running signalp locally...
## 2 sequences to be truncated
## Ok for single processing
## Number of submitted sequences... 100
## signalp &lt; 4, calling parser for the output...
## Number of candidate sequences with signal peptides... 8</code></pre></div>
<p>The above code under the hood will run <strong>signalp-2.0</strong> prediction and output result as an instance of <code>SignalpResult</code> class.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">class</span>(step1_sp2)</code></pre></div>
<pre><code>## [1] &quot;SignalpResult&quot;
## attr(,&quot;package&quot;)
## [1] &quot;SecretSanta&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">slotNames</span>(step1_sp2)</code></pre></div>
<pre><code>## [1] &quot;mature_fasta&quot; &quot;sp_version&quot;   &quot;sp_tibble&quot;    &quot;in_fasta&quot;    
## [5] &quot;out_fasta&quot;</code></pre>
<p><code>SignalpResult</code> object contains 5 slots:</p>
<ul>
<li><p><code>in_fasta</code> - original set of input amino acid sequences;</p></li>
<li><p><code>sp_tibble</code> - parsed <strong>signalp</strong> tabular output for positive candidates;</p></li>
<li><p><code>out_fasta</code> - full length amino acid sequences of positive candidates;</p></li>
<li><p><code>mature_fasta</code> - mature sequence for the candidate secreted proteins, i.e sequences with removed N-terminal signal peptides;</p></li>
<li><p><code>sp_version</code> - version of <strong>signalp</strong> used.</p></li>
</ul>
<p>You can use accessor methods to access contents of individual slots:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">getInfasta</span>(step1_sp2)</code></pre></div>
<pre><code>##   A AAStringSet instance of length 100
##       width seq                                        names               
##   [1]   411 MSTDIQTTQPHKKASSRRHQ...KDGLEKFCKGECPDICNKS ALI_PLTG_87
##   [2]   275 MATNPPTVAPVELRELNFAF...NRAGGYAAGRLGGVDIDSV ALI_PLTG_47
##   [3]   298 MDGEIQGLMVSGIGIAAGSN...DLIWLLLSACWQRMALEHP ALI_PLTG_86
##   [4]   172 MRSYLAKHPRYPKAAPTKAT...HIRTRLAYASTMLKYSNKD ALI_PLTG_91
##   [5]   340 MENGAEECLAYSRAKLGHIA...HFHAFDKASRNTIQATARR ALI_PLTG_35
##   ...   ... ...
##  [96]   240 MTRQGYLIVHDKRSRPTVRY...VTPHYVPTHFGRGNNQEAH ALI_PLTG_2
##  [97]   650 MQRLVQRDKDAPVVREGDDD...DRLSLNALLEAFPPMNSRL ALI_PLTG_96
##  [98]   318 MKGLRQQSKPNAQDERGGYG...KGSDEAVLQFVARYRANHL ALI_PLTG_24
##  [99]   121 MHFNALKRAAELAGSPDIIT...GNAFVLKVSAKSRLMYDLR ALI_PLTG_57
## [100]   897 MFGRASANAVEASGSAEPKT...TVNPPLSLTARVNIELDTK ALI_PLTG_52</code></pre>
<p>Please note, that by default <code>signlap</code> truncates all the input sequnces longer than 2000 a.a, in this case ’_truncated’ is added to the original sequence ids to keep track of the changes. Alternative option - is discard long sequences completely before the analysis, to do so set <code>truncate = FALSE</code> when running <code>signalp</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">getSPtibble</span>(step1_sp2)</code></pre></div>
<pre><code>## # A tibble: 8 x 9
##       gene_id  Cmax  Cpos  Ymax  Ypos  Smax  Spos  Smean     Prediction
##        &lt;fctr&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;fctr&gt;         &lt;fctr&gt;
## 1 ALI_PLTG_81    22    33 0.388    33 0.826    25  0.471 Signal peptide
## 2 ALI_PLTG_23    22    22 0.683    22 0.969     8  0.803 Signal peptide
## 3 ALI_PLTG_38    22    20 0.696    20 0.994     9  0.954 Signal peptide
## 4  ALI_PLTG_5    22    25 0.687    25 0.906     3  0.637 Signal peptide
## 5 ALI_PLTG_72    22    22 0.804    22 0.986     8  0.805 Signal peptide
## 6 ALI_PLTG_73    22    22 0.316    22 0.927     3  0.721 Signal peptide
## 7 ALI_PLTG_83    22    34 0.506    20 0.956    13  0.846 Signal peptide
## 8 ALI_PLTG_59    22    17 0.288    19 0.948     1  0.666 Signal peptide</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">getOutfasta</span>(step1_sp2)</code></pre></div>
<pre><code>##   A AAStringSet instance of length 8
##     width seq                                          names               
## [1]   133 MNFLKSKKTSTQATLLVSPLI...MGGLMPNLHVDLENGIGNYP ALI_PLTG_81
## [2]   653 MKIVALVTFCIATLDSSIVFA...SEDPVYPLVKEYSDVVSKHP ALI_PLTG_23
## [3]   495 MVVRVLLVVLALLAVGVQSKA...FAIVTIGASDGRVRYMNPPS ALI_PLTG_38
## [4]   210 MKLSSVCTIIFGLVFIDFNNV...SDGSVTTAAQKAAALVKANA ALI_PLTG_5
## [5]   118 MRLKFSILIFGAVLLATTTNA...NIRYASMLQDFLNTYHRRGV ALI_PLTG_72
## [6]   154 MYLKFAFMATVATAVVNIAIG...YAKLLLAYLNDFKRTSGLRS ALI_PLTG_73
## [7]   164 MVSTSRVFALLLLPSPSARIF...SMVEHIKTTKRVIDEVQDHV ALI_PLTG_83
## [8]   125 MLMSTPAALMLRLAQSSLKPT...DQGAHVATSGTSNNVLLWHK ALI_PLTG_59</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">getMatfasta</span>(step1_sp2)</code></pre></div>
<pre><code>##   A AAStringSet instance of length 8
##     width seq                                          names               
## [1]   101 YGPPADVRPMVVRMKSDYRSY...MGGLMPNLHVDLENGIGNYP ALI_PLTG_81
## [2]   632 AECTVDELTEISTIYSEAMTD...SEDPVYPLVKEYSDVVSKHP ALI_PLTG_23
## [3]   476 KATRVRKSWTAYTSDEKEIYL...FAIVTIGASDGRVRYMNPPS ALI_PLTG_38
## [4]   186 QCSDVHVVFARGSGEAAGLGI...SDGSVTTAAQKAAALVKANA ALI_PLTG_5
## [5]    97 DDNSNKRLLRRQEKTDTTNGD...NIRYASMLQDFLNTYHRRGV ALI_PLTG_72
## [6]   133 NVIDLPMPLTLESPNFVESAA...YAKLLLAYLNDFKRTSGLRS ALI_PLTG_73
## [7]   131 ITTKLKQFNGTGFPAWGVQDK...SMVEHIKTTKRVIDEVQDHV ALI_PLTG_83
## [8]   109 SLKPTVGKAITVGGNHQIPTK...DQGAHVATSGTSNNVLLWHK ALI_PLTG_59</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">getSPversion</span>(step1_sp2)</code></pre></div>
<pre><code>## [1] 2</code></pre>
<p>Next, imaging we would like to run all the <strong>signalp</strong> version on the same input for comparison. To do so, simply change <code>version</code> parameter:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">step1_sp3 &lt;-<span class="st"> </span><span class="kw">signalp</span>(inp,
                     <span class="dt">version =</span> <span class="dv">3</span>,
                     <span class="dt">organism_type =</span> <span class="st">'euk'</span>, 
                     <span class="dt">run_mode =</span> <span class="st">&quot;starter&quot;</span>,
                     <span class="dt">paths =</span> santa_paths)
## Version used... signalp3
## running signalp locally...
## 2 sequences to be truncated
## Ok for single processing
## Number of submitted sequences... 100
## signalp &lt; 4, calling parser for the output...
## Number of candidate sequences with signal peptides... 8

step1_sp4 &lt;-<span class="st"> </span><span class="kw">signalp</span>(inp,
                     <span class="dt">version =</span> <span class="dv">4</span>,
                     <span class="dt">organism_type =</span> <span class="st">'euk'</span>, 
                     <span class="dt">run_mode =</span> <span class="st">&quot;starter&quot;</span>,
                     <span class="dt">paths =</span> santa_paths)
## Version used... signalp4
## running signalp locally...
## 2 sequences to be truncated
## Ok for single processing
## Number of submitted sequences... 100
## Number of candidate sequences with signal peptides... 5</code></pre></div>
<p>In this case <code>signalp-4.1</code> version resulted in fewer number of candidates. Please note, taht despite differences in the output format generated by multiple versions of <strong>signalp</strong>, <code>signalp</code> returns <code>sp_tibble</code> in a standard format.</p>
<p>In order to pipe result from differenr versions of <strong>signalp</strong> just switch to the <code>run_mode = 'piper'</code> for the second and other downstream steps. In this case <code>signalp</code> will run the anlysis using contents of <code>out_fasta</code> slot as an input. Say, we want to do the following piping: <strong>signalp2</strong> -&gt; <strong>signalp3</strong> -&gt; <strong>signalp4</strong>. We will re-use <code>step1_sp2</code> object generated earlier in the <code>starter</code> mode and use it as an input for the next step:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">step2_sp3 &lt;-<span class="st"> </span><span class="kw">signalp</span>(step1_sp2,
                     <span class="dt">version =</span> <span class="dv">3</span>,
                     <span class="dt">organism_type =</span> <span class="st">'euk'</span>, 
                     <span class="dt">run_mode =</span> <span class="st">&quot;piper&quot;</span>,
                     <span class="dt">paths =</span> santa_paths)
## Version used... signalp3
## running signalp locally...
## Ok for single processing
## Number of submitted sequences... 8
## signalp &lt; 4, calling parser for the output...
## Number of candidate sequences with signal peptides... 8</code></pre></div>
<p>Similar with the <strong>signalp3</strong> -&gt; <strong>signalp4</strong> pipig:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">step3_sp4 &lt;-<span class="st"> </span><span class="kw">signalp</span>(step2_sp3,
                     <span class="dt">version =</span> <span class="dv">4</span>,
                     <span class="dt">organism_type =</span> <span class="st">'euk'</span>, 
                     <span class="dt">run_mode =</span> <span class="st">&quot;piper&quot;</span>,
                     <span class="dt">paths =</span> santa_paths)
## Version used... signalp4
## running signalP locally...
## Ok for single processing
## Number of submitted sequences... 8
## Number of candidate sequences with signal peptides... 5</code></pre></div>
<p>With the input fasta files containig more than 1000 sequences,<code>signalp</code> will automatically switch parallel mode: will split the input into smaller chunks and will run prediction as an embarassingly parallel proces using all the CPUs available. Accurate execution time depend on the number if CPUs available and the input file size. With 48 CPUs it takes ~2 minutes to run <code>signalp</code> on the input fasta file with more than 40’000 seqeunces.</p>
</div>
<div id="tmhmm" class="section level2">
<h2>TMHMM</h2>
<p><strong>TMHMM</strong> predicts transmembrane a-helixes and helps to identify integral membrane proteins based on HMMs [<span class="citation">Krogh, Bjorn Larsson, and Sonnhammer (2001)</span>; Sonnhammer1998]. It is important to exclude proteins with trans membrane domains locates after signal peptide, as they will be retained in the membrane. Quite often <strong>TMHMM</strong> is not able to distinguish between N-terminal signal peptides and transmembrane domain. So, it is better to run it on the mature sequences with signal peptides removed. <code>tmhmm</code> function can handle input objects of CBSResult superclass having <code>mature_fasta</code> slot.</p>
<p>Here we will use output of the <code>signalp</code> function as an iput:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tm &lt;-<span class="st"> </span><span class="kw">tmhmm</span>(step1_sp2, <span class="dt">paths =</span> santa_paths, <span class="dt">TM =</span> <span class="dv">0</span>)
## running TMHMM locally...
## Number of submitted sequences... 8
## Number of candidate sequences with signal peptides and 0 TM domains in mature sequence... 8</code></pre></div>
</div>
<div id="targetp" class="section level2">
<h2>Targetp</h2>
<p><strong>Targetp</strong> predicts subsellular localisation of secreted eukaryotic proteins based on the presence of signal peptide (SP), chloroplast transit peptide (cTP) or mitochondrial targeting peptide (mTP) in the N-terminus <span class="citation">(Olof Emanuelsson and Heijne 2008)</span>.</p>
</div>
<div id="wolfpsort" class="section level2">
<h2>WolfPsort</h2>
</div>
<div id="check-for-terminal-er-retention-motifs" class="section level2">
<h2>Check for terminal ER-retention motifs</h2>
</div>
<div id="m-slicer" class="section level2">
<h2>M-slicer</h2>
</div>
<div id="pipers-andor-starters" class="section level2">
<h2>Pipers and/or starters</h2>
<p>Here is just a short summary indicating wich fucntions can work in piper and/or starter mode.</p>
<p>(add table here)</p>
</div>
<div id="build-pipelines-simple-workflow" class="section level2">
<h2>Build pipelines, simple workflow</h2>
</div>
<div id="build-pipelines-stringent-workflow" class="section level2">
<h2>Build pipelines, stringent workflow</h2>
<p>Note the various macros within the <code>vignette</code> section of the metadata block above. These are required in order to instruct R how to build the vignette. Note that you should change the <code>title</code> field and the <code>\VignetteIndexEntry</code> to match the title of your vignette.</p>
</div>
<div id="styles" class="section level2">
<h2>Styles</h2>
<p>The <code>html_vignette</code> template includes a basic CSS theme. To override this theme you can specify your own CSS in the document metadata as follows:</p>
<pre><code>output: 
  rmarkdown::html_vignette:
    css: mystyles.css</code></pre>
</div>
<div id="figures" class="section level2">
<h2>Figures</h2>
<p>The figure sizes have been customised so that you can easily put two images side-by-side.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(<span class="dv">1</span>:<span class="dv">10</span>)
<span class="kw">plot</span>(<span class="dv">10</span>:<span class="dv">1</span>)</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAC61BMVEUAAAABAQECAgIDAwMEBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4PDw8QEBARERESEhITExMUFBQVFRUWFhYXFxcYGBgZGRkaGhobGxscHBwdHR0eHh4fHx8gICAhISEiIiIjIyMkJCQlJSUmJiYnJycoKCgpKSkqKiosLCwtLS0uLi4vLy8wMDAxMTEyMjIzMzM0NDQ1NTU2NjY3Nzc4ODg5OTk6Ojo7Ozs8PDw9PT0+Pj4/Pz9AQEBBQUFCQkJDQ0NERERFRUVGRkZHR0dISEhJSUlKSkpLS0tMTExNTU1OTk5PT09QUFBSUlJTU1NUVFRVVVVWVlZXV1dYWFhZWVlaWlpbW1tcXFxdXV1eXl5fX19gYGBhYWFiYmJjY2NkZGRlZWVmZmZnZ2doaGhpaWlqampra2tsbGxtbW1ubm5vb29wcHBxcXFycnJzc3N0dHR1dXV2dnZ3d3d4eHh5eXl6enp8fHx9fX1+fn5/f3+AgICBgYGCgoKDg4OEhISFhYWGhoaHh4eIiIiJiYmKioqLi4uMjIyNjY2Ojo6QkJCRkZGTk5OUlJSVlZWWlpaXl5eYmJiZmZmampqbm5ucnJydnZ2enp6fn5+goKChoaGioqKjo6OkpKSlpaWmpqanp6eoqKipqamqqqqrq6usrKytra2urq6vr6+wsLCxsbGysrKzs7O0tLS1tbW2tra3t7e4uLi5ubm6urq7u7u8vLy9vb2+vr6/v7/AwMDBwcHCwsLDw8PExMTFxcXGxsbHx8fJycnKysrLy8vMzMzNzc3Ozs7Pz8/Q0NDR0dHT09PU1NTV1dXW1tbX19fY2NjZ2dna2trb29vc3Nzd3d3e3t7f39/g4ODh4eHi4uLj4+Pk5OTl5eXm5ubn5+fo6Ojp6enq6urr6+vs7Ozt7e3u7u7v7+/w8PDx8fHy8vLz8/P09PT19fX29vb39/f4+Pj5+fn6+vr7+/v8/Pz9/f3+/v7///+sPjywAAAACXBIWXMAAA7DAAAOwwHHb6hkAAANTElEQVR4nO2de1wVVR7AD2JXQQR5LPJS8xlGtKLm7vpIkHVT0UqpVbRSg/KNoGkP1LS1xAfmW3NJyAdmksSumq9NU7NNdNXEZdvSHqtICgoC954/d2YukHZn5ndnPGdmlN/342fO9c5vzhm/zpw5c+ecM4QiqhCzd8DqoCAAFASAggBQEAAKAkBBACgIAAUBoCAAFASAggBQEAAKAkBBACgIAAUBoCAAFASAggBQEAAKAkBBACgIAAUBoCAAFASAggBQEAAKAkBBACgIAAUBoCAAFASAggBQEAAKAkBBACgIAAUBoCAAFASAggBQEAAKAkBBACgIAAUBoCAAFASAggBQEAAKAkBBACgIAAUBoCAAFASAggBQEAAKAkBBACgIAAUBoCAAFASAggBQEAAKAkBBACgIAAUBoCAAFASAggBQEABjQWndrU505KNi0udHcwT9buMJS5MXExjl+8fdJ050LjJJ0FG2+THmcthaO616tVsNjUZBcixMkZI++QYLKkqsxyeXRX7cGJUjJa/PN1hQ6bZ6mq9ikR83xq2RktTFpp1iPuvY5seYD/rZhWV525MoSJ7a+EGHLxXGTKYoSIGarF5h/bdSFASCggBQEAAKkqh4a9DAjDK5NShI5L/txxbunhh+RmYVChIZsVBcromVWYWCBBze18Wk2ve66zoUJFDZ3JlGfOe6DgWJtDktLi8F1LiuQkEimX1/ovTq4FdlVqEgEfvrwcOeDp4mcwChoDq+z98hUwFRFASCggC4Cio7c0tplcmCij8+obhrd8JJUP6TT+TQ1TZim1MrH2CqoEsJbYZ267zPrVg+graQngkPpLdcUDjLc7V8hJmCarvPFa5XnwafcyeYj6BHRzvoZiLe4EyOlo8wU9D+7lIyb5o7wXwEeeUIFRA5KHza7H3798fj6/FcrCU/tqycICW7B7oTzEdQ51mUHiLiQfJml9u/r9hbj9daLfmx5f0/S8nm4e4E8xH0judLbwT3Cd1zebuvXPOdmnuKXQoS24T2uPfcCeYjqCYjLGi6I4UQMqRCPsLUq1hW27X/3NX/T3J3Fi5wawc5hD8Hs084FFab2w46NrLbkA12t0KxJQ2AggBQEEBjEFT1Tr+uicd1btwIBFXEJB48sz5CZ8OrEQj6yyhxeSFQ9rkgSCMQ9PheKXmiQNfWjUBQT2f18+wWXVs3AkHPLxOXNQ+e1rW1DkGOa6VK7WP3MVDQVyH7hZo6ZYi+rbUKKp7VqRkhtg6vXNBXXj1GXub3dHokLmjsz/o21ijouG9YyrL3s7MmtvP7Ul+BdRjaUKwp2ufugAIXNArq/3i5869VT8bpLVLifm1J+66v//suv7sq934V1DOpvn6e/hi4Sc357xVrc56C/pWTd5FZZhoFbfMYsO7ImbOfv5fQJE8luHbhcFqT4UVIRI5CBD9B5SPbJY0Innv3F1onWq9iBbFEIla1XTqPzKBzmk7/+MMk8qF8BD9Bz42tpPTHx1gNddDeDrpStGfPySvqwW0mUho6R/z0Qg/5CG6CfvaTLiNHFB43aYZPS9p/A621FYqfNvnc/v1nDeP5mizSkp8Gipxmbnoxyo+PoIQhtbTvK+KnMXfU5dVf1o/o8+Z1BH0bItU+JeGM8uMj6FRgv9xc77S9u0YShVqaXx0U84G4nDKZUXYaBaXfhlr0ueSWUl0erTRujp+gr0In5W8d+turjLLTKGiGPwnoWId6fNWFwwXHvjWjHXR13rBnVlWzyk3rKXbeM5NJufdrS5rSP6AgeeoFHTrFpNz7VxAjmAmqfTcmIGa5W4/ZdaFdUJnzB49yfQ8J6mEm6Kn4o6VHBw5ldevlgnZBZICUDLi7Q4qVoL8/Kh48Nd0K2WTninZB4519wxaPv6tyWQmavUBK3p7JJjtX7vU6KG2JlGRNZZOdK/e6oGznw4phG9lk58q9Lqiya0YVrZobWckmO1fudUH0UqJvtN8I+YEoLOB0swrCsKF4regas7xc4XazCnDftqTxZlUJPTer1R9dVlqlW9DpVZkHdG6qC643q2VE8d+iU5B9SsTLM6JjFb2zh89VLMlJIolPSpKP0CloSaxwJ2if8aSujXXBR1Ac8ekt0ItE9e4tH6FTUJTUGepWwP90ba0HPoLsK32eK5U5xQ761+Pxjpb8GmjpnBuh+wldW+uBV0OxJDYk31WQ42o9LfQdQR2l+Uccrfk1DH8Nt5a0fUWL0ReYV9KzR4k//KxSOG95wPFWoyTWn7mgG/G9lm94psPXujbWBc97MfuGdMWOenrbQY4dk8evvKlvW13c8zervEFBAJYXxOwRqU6sLciR/XDzgDHGXdNlsLag13ocrb3yZsT3bMvWhKUF/RAg9WR7ZTrbsjVhaUH5CVJyFO5Qyw9LC/pomJQcV+jnaAiWFvRdkHRvOmcS27I1YWlBdHLseVq9IuQbtmVrwtqCapYGt24x2K1pbHhhbUECP7o5UxQvLC/IbFAQAC9BNSU3pLTiB/n18oLOLZq+WufIQF7wEVQ7z4vYUsX7zEyF7WQFLQxJW5oUtl9LSdzhIyjLY2peqqf4xEeLoH0dxYcVB0NlJi02Dz6CHhabdhvJDm2Cxq2QkgS1kWiGw2mSt83icniHyl8J2k8aWO661SBnR8Npy7QUxRs+gh5KFZfftRrnUDqC5N4ONf5dKUnYrqUo3vARtNgjZedNSreS8RM0CDrQXhy8va8x1EE1C1uRs0K6LZhoEEQXtU5dMjL8oJaSuMOrHeS4KPUavPWpwrQ98i9gK16cts5Sx495LWmLv6HuF1AQgLmCrlrsvkIGMwXtfMjfL+pvbPNljomCcjsdonRP23y2GbPGREHtpd5inz7CNmPWmCfocqD0we5lZF8N7ZgnqLSVNAiutnkV25wZY+Ip9vA+8cPOXmwzZo2JggrDttyoyA45xDZj1pgmaPaaNbO6NGve9bU1t7M8fjTE8MFgyOCnwZABI6GIpMHSHoWbJGhdikhycsqdjLB1hQjzA0NahYIhtg5QRJem0h5NdXcoEWNBChR3AkM2vgCGvAg/Mok8C0VcDgIzuQMUBICCAFAQAAoCQEEAKAgABQEYI+g/kWBIzotgyMt/BUOiiqGIshAwkzswRhAFZrAUqIabttfhTldwOe6E3I5Bgu5dUBAACgJAQQAoCAAFAaAgABQEgIIAUBAACgJAQQBGCKpd0tU7arnCC9gbsMcpzElUx6mhAW3fVH2Jak1mF+9uW1UCZjhnsMvr5Rf3BbA3DRghaKHH1A8meMwDolYSVUHnggavn0DeVguZ4zlzUxLZpbi+OEASVNBk1Lq+fuDvInUYIMjhnywsp3ipz31c4uOjKii5n3AIpqu+6DrsJaGwHk8prD3UpymRBMXGOej10Fmqe/MLBgi6SHYKyzxSohZkjx3TW01QVctcCpykjqA3hGXCEwqrT2VmSkfQFSL20E2Bf+N0YoCgyrNiV6FU2w21oBUhpaqC/k1y+9rC56oehamt//Hzlgc2KQd0FAUVkSPCclkzN6eqNuoqttFzmtrqkhY7qaqgw8R7xicZTTPUMrH3IYSkqQRIgvYQcbqr94mbD+eNEXRxBBml9nupPVaQoyqogIjzNKR6qxxCjpdC1hx4q+Vq5Yg6QeIP+9nEzZ9eDRG0rVWHHaoBG/wvlJX1SixTnhPmGBFfq/oxOa+cySGyW1gusCn3QZYEnSSfC8ssm4VOsTwyDpj2OL1uMNVHihHfSpfvD4nKhCjZRJyoslA6g+Tp6KykxUmqJ3VQ36MGDBB0K3g89L914YBAVPwB5ak4HdEjheXoCJWcvpBeejbbpnwuS4Jo/wThovigu5PBGyBoH5mQKVIBxKnWQTSXPJczTumdSxKOQb4LtkzxVKnHnYIKPNIKnvJz932pBghaW3f+KAyUbkBdEN3U0ydGfVRn+ezOXlErVFpLTkF0aw/fWLentMSbVQAUBICCAFAQAAoCQEEAKAgABQGgIAAUBICCAFAQAAoCQEEAKAgABQGgIAAUBICCAFAQAAoCQEEAKAgABQGgIAAUBGA9QatcnlGvCjRlR5ygIAAUBGBNQe02ZXb1TSyl9HpyePikLEGQY32M9yObxamr91Ka1/SkYbtjUUG9h32y3DaROvp7z9/4ez9B0BKPyZtHi92DxnauvBb2unG7Y1FBkbWUPv8YPUC2U1rZJpCW+4kdv5/tROmV38ydFgn06GOJRQWJHeRmdaeZLcXBGemB9Bg5LXzIb1JNaa6t6REDd8eighZRSVD6Q+I3SwPp9rpeat9QWh0UqTqkhTEWFZRJJUFLpSNoZiA9THZ/JXKD0rdbNzPy9S/WFvSZWAdVtQukV73eE75ZPdZBv262NcP/J+N2x9qCHAO852f3aS9cxd7wnrNtZpP51N53oONm+zHG7Y61BdHy5IjQCZ+I7aBlUV6Ryxx0ZbNicVyCce/BsZ4gi4GCAFAQAAoCQEEAKAgABQGgIAAUBICCAFAQAAoCQEEAKAgABQGgIAAUBICCAFAQAAoCQEEAKAgABQGgIAAUBPB/+8gEMS55W1gAAAAASUVORK5CYII=" /><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAASAAAAEgCAMAAAAjXV6yAAAC9FBMVEUAAAABAQECAgIDAwMEBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4PDw8QEBARERESEhITExMUFBQVFRUWFhYXFxcYGBgZGRkaGhobGxscHBwdHR0eHh4fHx8gICAhISEiIiIjIyMkJCQlJSUmJiYnJycoKCgpKSkqKiorKyssLCwtLS0uLi4vLy8wMDAxMTEyMjIzMzM0NDQ1NTU2NjY3Nzc4ODg5OTk6Ojo7Ozs8PDw9PT0+Pj4/Pz9AQEBBQUFCQkJDQ0NERERGRkZHR0dISEhJSUlKSkpLS0tMTExNTU1OTk5PT09QUFBRUVFSUlJTU1NUVFRVVVVWVlZXV1dZWVlaWlpbW1tcXFxdXV1eXl5fX19gYGBhYWFiYmJjY2NkZGRlZWVmZmZnZ2doaGhpaWlqampra2tsbGxtbW1ubm5vb29wcHBxcXFycnJzc3N0dHR1dXV2dnZ3d3d4eHh5eXl7e3t8fHx9fX1+fn5/f3+AgICBgYGCgoKDg4OEhISFhYWGhoaHh4eIiIiJiYmKioqLi4uMjIyNjY2Ojo6Pj4+QkJCRkZGSkpKTk5OUlJSVlZWWlpaXl5eYmJiZmZmampqbm5ucnJydnZ2enp6fn5+goKChoaGioqKjo6OkpKSlpaWmpqanp6eoqKipqamqqqqrq6usrKytra2urq6vr6+wsLCxsbGysrKzs7O0tLS1tbW2tra4uLi5ubm6urq7u7u8vLy9vb2+vr6/v7/AwMDBwcHCwsLDw8PExMTFxcXGxsbHx8fIyMjJycnKysrLy8vMzMzNzc3Ozs7Pz8/Q0NDR0dHS0tLT09PU1NTV1dXW1tbX19fY2NjZ2dna2trb29vc3Nzd3d3e3t7f39/g4ODh4eHi4uLj4+Pk5OTl5eXm5ubn5+fo6Ojp6enq6urr6+vs7Ozt7e3u7u7v7+/w8PDx8fHy8vLz8/P09PT19fX29vb39/f4+Pj5+fn6+vr7+/v8/Pz9/f3+/v7///9VMu2zAAAACXBIWXMAAA7DAAAOwwHHb6hkAAANNElEQVR4nO2de1wVVR7AD2LXuIKIKAKGJKKSsWy+TcUENR/s1q66IvgAFVPzia82UdPygZro+n4QKpliPhI1NZFNK5JSl1JTl8rMsiRMTBC4c/7ZmblgtszMby73zNy58Pt++My53vndc87n6zzOzDlzhlBEEeLoChgdFASAggBQEAAKAkBBACgIAAUBoCAAFASAggBQEAAKAkBBACgIAAUBoCAAFASAggBQEAAKAkBBACgIAAUBoCAAFASAggBQEAAKAkBBACgIAAUBoCAAFASAggBQEAAKAkBBACgIAAUBoCAAFASAggBQEAAKAkBBACgIAAUBoCAAFASAggBQEAAKAkBBACgIAAUBoCAAFASAggBQEAAKAkBBACgIAAUBoCAAFASAggBQEAAKAkBBACgIAAUBoCAAFASAggBQEAAKAmAsaHoHnvZt27bvYHB6/OgYQV1Tc3NnebUK8p6ba2xaXXCQoI9pSocrlH7x9Da2GbMmzGGCLH4XhQ85wWwzZo2+gi4MqcQ9/ZaP9bv6v7HIWTP0FVSwp5LH19/xsAhflTxeyiJnzXDULua+mXbaK3xI7c02Y9Y4UNDHPslXv1rkc55txqxxoCB6bXhw6/hv2ebLHEcKcgpQEAAKAjCWoIyhPRPUVkgnjCTIMrhTevZK/w1sS7ITIwlK7ya0Gb9ufJNtUfZhJEHRadZkO9ui7MNIggYcFpPJKWyLsg8jCZqyUEy6H2VblH0YSdBFnzP8kXrF04a6ejWSIHokMHxYSM98tiXZiaEE0ftZ6Z+zLcdujCXIgKAgAE0FFV58ILdKraCvM3PuV6Nkdmgk6OAL/XbSDSZimlcuHaBOUEG038DOzQ/YVDRjtBH0DukU9Viix6Ijs11lrqzUCeo7qZjSHP+PbSmbMdoI+nMsR3eRxfynl8OkI1QJ+uJJcftbP8yWshmjjSC3nfwBiJziP+0yP/r9p30qcV2uIpu9g8Tk3DO2lM0YbQS1mk1pNhE2ktdaP/r9veOVuG1Skc3RSDH5INyWshmjjaClruPm+vTwO/bz3gavSEeo2sWKmnwpJP9YYkvZjNFGUFmSf+NpXAIhZOA96Qh1B+mdfim5R6O6OrLvVbN2EMf/nUrL5WRWq2wH5Y3s0HeNQy9esSUNgIIAUBCAswm6M6dL2KirbOuiiJMJ+i5w4icXlvscZ1sZJZxM0Kj5wvJkkNzJkT1OJsjfOhgk6ArLuijiZII874hJO/1uzDqZoGePCMsCzyKmlVHCyQTtC/4PpT8NmMa2Mko4mSCa5t/xuUazdbz4cDZB9P7Z7F+Y1gTARkGJj2BXuTW1JT3DizRqWYFd5dZUQfSKa7LqvMuu3JRt0LET9ODE1hOyvUv2Y/MxqJsaQeWLB9GyJDdCntgpE8FM0Eetw+PDW51mlFtVbBaUnacieAGZQefVnfbeuzHkXekIVoJ+aioMKjrS9Bab7KqizVksYAKlfvOET6M6SkewErQ6TkzGrGKTXVVsF1RobcUWFSoEe22l5Sax0bvD/dHvP3z4IF+dZTZWVIYpVjP/msgmu6rYLohYnz7prbRJRQ0sp+GzhE/DOz/6felnlQ/ymRltQfOtvSZz57LJriq2Cxpt7fJbPlohOM+7Z3q6efrxQ9FE5ijNahf7POA2vyxo/hmb7KqiUUv68lgPIhCWLhPA7Cy2oPnqo6sD5zHKrSrVEMT9WqDiflXJtdOZOdd1aAfljHt+nIajG2wVdHV2cD1CTEGzrtlXbk1tSX/awD9h1fa0lAmBnvbt9TVVUK/nKm5VlbwQaVe5NVVQgy2V/z7kaVe5Ggo6O9CvRdx3rHKzUVCnmMrD7rTO8rEq0E7Qfv+3buYv9P2KUXY2Ctrj0nvzmYuXPtoWVSfDrnI1E2QJ+ERI3hzMKD9bz2KZEWL7hkRk2leuZoKuBInJLW9G+dneDrp94dix87ftLVczQXlPi8lddyBOLU53TxqiuOENITnAathejRNEX+vGt2GzA44xyq7mCeLe9G4T2PIQq+xqYq9G2eVv2GWGvRoAWvZqKFFjBanr1aigdP/Pcqt0FfTtlqWHZB6qAdGmV6OCQpIlt0pPQct84meFh12u3o+1OYvFWBlC+sTESEfoKOhQmx/45YbQ6m1D2giKJO7debqQ0O7dpSN0FPQ3623xZz6q1q+1EWRZ5z6iQGIXO+VVictSW/Kziw65YhL9drV+rVVDMT/C92BVQdwvldTXbwvqb31isccH1fq1Zi1py9r6sdcMcZDe1q2EX54IKKnWrzW81MiP8DKEIEtc6+TtCb7Z1fu1ltdilq2Jsn0furaDsmaMSC6o5m9r3sUqY1AQQG0VpHqcbK0UVL42uF6Tl2SvE/9ArRQ0ptc5y4+Jwb+qia2Ngr5sJk4HMmqxmuDaKGizddjewSg1wbVR0MYxYpLZX01wbRR07klxXPWEBWqCa6MgOvSF67T4jQBVvZ+1UlDxAi/f+kPUzWRdKwXx/KC2pVhbBakGBQFoJags3zply70fpNcbTdDRV185KDkkVxtB5QvciGmqsJsny/zOWIKKnm+38I2u3aROa9oISnGZnDHVVejxcQ5BL8dZKOWmR0us0kZQW+HZklSyz1kEeYlzfBd5FFddpdEkb7uE5aCg4v8TdJI8ZI0t+WnMfTdrGnC96jptBLWZKiy/axjPyW1BXR05N2IVGn0vLIs8JCb91EbQcpeEA3xhu8no8U4haNLIcv4YNEWqm1yjSd4WNySX+HSPD3EKQff6h81/vXO4VM+HVu0g7oZ4wHtwQma+RGMJovR40j8zdWwHwRhNkCwoCAAFyVH2vbjLoSBpbsaY/d1n3EVBMhQFz/2N/jQikkNB0qSIl2XcM8dRkDSx1tcMvfI6CpJm5DYxmbkEBUmzSexULA05g4KkKWk35lsur98QB57F5myUYk2fWIhBA8CQAX8HQ3pHK68f2tbs0mTwuo0bmzlI0OYESQabnoLw9wRDGvqBIaYgKKJ1XbFGk1WNAWEvSIar8LvDU0eBIWPg25Uhl6CInxuDmfwBFASAggBQEAAKAkBBACgIAAUB6CPo6xAwZOcYMOSlt8CQUPDNCoW+YCZ/QB9BFB4WVwo3be/Cc7uqGH5n4/wkOglyXlAQAAoCQEEAKAgABQGgIAAUBICCAFAQAAoCQEEAeggqX/GUOXQNNM+PJVJmTqIK8v7SqPlrFqWIsuTW5na7FQJmWGewy+jiGXkWqM1D9BC02GXy2+NdoAcB1xFFQZcbD9gynii+/Xie68wdMUR+JsGrjURBmXWGbQ73VPvGSR0EcV5j+eUktzLFqHx3d0VBY3vym2DiIKUQ/3F8YR1flFmb3aMuEQVFRHL0rt9sxdr8jg6CbhBhpp8Mkq8UZIkY3l1JUIlHOgV2Uq6x8JKJqH4yq/OSk8Ut6DYRRugmwPc4reggqPiSMNJ9qknx3ddrfQsUBf2XpIebms1X3AqnNv33nXce2yEf0FIQdIGc4Zer6ql8I6deZ7FU1ylKq/PrH6CKgk4T84zDSXWTlDKx9CCETFcIEAUdIxf55XaisnNeH0E3BpNhSvdLLRG8HEVBmUR4AeJUs8ImxI3z3Zj1uscG+YgKQcKN/TSi8tarLoL2NAzapxiw1etaYWGXIYXyj+TmEOE1vu8RhTexZpP3+eUi0x3ZCFHQeSLMm5diMtAulkHiJZ7YepTEioep9stGXBdP3++Sm/KZpBFh4pcj4h4kTUvrQTqVX04MUq7RQ3QQ9MBnNPS/dS2LJ7RPlvzUNlyYMEA19gmFnM6KLz2bY5Lfl0VBtFcUf1J8ciZQpUp0EPQBGZ8scA+IUzwG0XQyYme83DuXRLj+DRa9M8k1ST7CKijTZXrmi55qX0Gjg6BNFfuPzIPSD1EWRHd0cm+v/MqGojmt3ELXKrSWrILo7o4NInKByjwEL1YBUBAACgJAQQAoCAAFAaAgABQEgIIAUBAACgJAQQAoCAAFAaAgABQEgIIAUBAACgJAQQAoCAAFAaAgABQEgIIAUBCA8QStr9JHvZ7VS+arAwoCQEEAxhQUuCP5qQZDCii9O7ZZs4kpvCBuS3vzn3YJU1cfpzSj7nndqmNQQd3/eniNaQLlepkXpj7ryQta4fLyrlhheFBcq+Jf/V/VrzoGFRRSTunIzjSL7KW0OMCbFnkKA7+HBlN6u8n8KSHAiD6WGFSQMEBudgea7CE8nJHoTXPIF/yHg3VKKU031T2jY3UMKmgZFQUlthG+WelN91aMUvuG0tLGIYqPtDDGoIKSqShopbgFzfSmp8n75wR+o3RJ03p6vnrB2II+FI5BJYHe9Be3bfw3G+I4+lW93Ulet/SrjrEFcb3NC9N6tODPYnPN8/bMrLOQWsL7cvdbDNevOsYWRIvGPuE3/rDQDloV6hayiqPr6l0Vnks4qVt1jCfIYKAgABQEgIIAUBAACgJAQQAoCAAFAaAgABQEgIIAUBAACgJAQQAoCAAFAaAgABQEgIIAUBAACgJAQQAoCAAFAfwPnWW7YdWET8QAAAAASUVORK5CYII=" /></p>
<p>You can enable figure captions by <code>fig_caption: yes</code> in YAML:</p>
<pre><code>output:
  rmarkdown::html_vignette:
    fig_caption: yes</code></pre>
<p>Then you can use the chunk option <code>fig.cap = &quot;Your figure caption.&quot;</code> in <strong>knitr</strong>.</p>
</div>
<div id="more-examples" class="section level2">
<h2>More Examples</h2>
<p>You can write math expressions, e.g. <span class="math inline">\(Y = X\beta + \epsilon\)</span>, footnotes<a href="#fn1" class="footnoteRef" id="fnref1"><sup>1</sup></a>, and tables, e.g. using <code>knitr::kable()</code>.</p>
<table>
<thead>
<tr class="header">
<th></th>
<th align="right">mpg</th>
<th align="right">cyl</th>
<th align="right">disp</th>
<th align="right">hp</th>
<th align="right">drat</th>
<th align="right">wt</th>
<th align="right">qsec</th>
<th align="right">vs</th>
<th align="right">am</th>
<th align="right">gear</th>
<th align="right">carb</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Mazda RX4</td>
<td align="right">21.0</td>
<td align="right">6</td>
<td align="right">160.0</td>
<td align="right">110</td>
<td align="right">3.90</td>
<td align="right">2.620</td>
<td align="right">16.46</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">4</td>
<td align="right">4</td>
</tr>
<tr class="even">
<td>Mazda RX4 Wag</td>
<td align="right">21.0</td>
<td align="right">6</td>
<td align="right">160.0</td>
<td align="right">110</td>
<td align="right">3.90</td>
<td align="right">2.875</td>
<td align="right">17.02</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">4</td>
<td align="right">4</td>
</tr>
<tr class="odd">
<td>Datsun 710</td>
<td align="right">22.8</td>
<td align="right">4</td>
<td align="right">108.0</td>
<td align="right">93</td>
<td align="right">3.85</td>
<td align="right">2.320</td>
<td align="right">18.61</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">4</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td>Hornet 4 Drive</td>
<td align="right">21.4</td>
<td align="right">6</td>
<td align="right">258.0</td>
<td align="right">110</td>
<td align="right">3.08</td>
<td align="right">3.215</td>
<td align="right">19.44</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">3</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td>Hornet Sportabout</td>
<td align="right">18.7</td>
<td align="right">8</td>
<td align="right">360.0</td>
<td align="right">175</td>
<td align="right">3.15</td>
<td align="right">3.440</td>
<td align="right">17.02</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">3</td>
<td align="right">2</td>
</tr>
<tr class="even">
<td>Valiant</td>
<td align="right">18.1</td>
<td align="right">6</td>
<td align="right">225.0</td>
<td align="right">105</td>
<td align="right">2.76</td>
<td align="right">3.460</td>
<td align="right">20.22</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">3</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td>Duster 360</td>
<td align="right">14.3</td>
<td align="right">8</td>
<td align="right">360.0</td>
<td align="right">245</td>
<td align="right">3.21</td>
<td align="right">3.570</td>
<td align="right">15.84</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">3</td>
<td align="right">4</td>
</tr>
<tr class="even">
<td>Merc 240D</td>
<td align="right">24.4</td>
<td align="right">4</td>
<td align="right">146.7</td>
<td align="right">62</td>
<td align="right">3.69</td>
<td align="right">3.190</td>
<td align="right">20.00</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">4</td>
<td align="right">2</td>
</tr>
<tr class="odd">
<td>Merc 230</td>
<td align="right">22.8</td>
<td align="right">4</td>
<td align="right">140.8</td>
<td align="right">95</td>
<td align="right">3.92</td>
<td align="right">3.150</td>
<td align="right">22.90</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">4</td>
<td align="right">2</td>
</tr>
<tr class="even">
<td>Merc 280</td>
<td align="right">19.2</td>
<td align="right">6</td>
<td align="right">167.6</td>
<td align="right">123</td>
<td align="right">3.92</td>
<td align="right">3.440</td>
<td align="right">18.30</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">4</td>
<td align="right">4</td>
</tr>
</tbody>
</table>
<div id="references" class="section level3 unnumbered">
<h3>References</h3>
<div id="refs" class="references">
<div id="ref-Bendtsen2004">
<p>Dyrl�v Bendtsen, Gunnar von Heijne annick, Henrik Nielsen, and S�ren Brunak. 2004. “Improved Prediction of Signal Peptides: SignalP 3.0.” <em>J. Mol. Biol.</em>, no. 340: 783–95.</p>
</div>
<div id="ref-Nielsen1997">
<p>Henrik Nielsen, S�ren Brunak, Jacob Engelbrecht, and Gunnar von Heijne. 1997. “Identification of Prokaryotic and Eukaryotic Signal Peptides and Prediction of Their Cleavage Sites.” <em>Protein Engineering</em> 10: 1–6.</p>
</div>
<div id="ref-Krogh2001">
<p>Krogh, Anders, Bjorn Larsson Gunnar von Heijne, and Erik L.L. Sonnhammer. 2001. “Predicting Transmembrane Protein Topology with a Hidden Markov Model: Application to Complete Genomes.” <em>J. Mol. Biol.</em>, no. 305: 567–80.</p>
</div>
<div id="ref-Nielsen1998">
<p>Nielsen, Henrik, and Anders Krogh. 1998. “Prediction of Signal Peptides and Signal Anchors by a Hidden Markov Model.” <em>Proceedings of the Sixth International Conference on Intelligent Systems for Molecular Biology (ISMB 6)</em>, 122–30.</p>
</div>
<div id="ref-Emanuelsson2008">
<p>Olof Emanuelsson, Soren Brunak, Henrik Nielsen, and Gunnar von Heijne. 2008. “Predicting Subcellular Localization of Proteins Based on Their N-Terminal Amino Acid Sequence.” <em>J. Mol. Biol.</em>, no. 300: 1005–16.</p>
</div>
<div id="ref-Sperschneider2015">
<p>Sperschneider J, Hane JK, Williams AH. 2015. “Evaluation of Secretion Prediction Highlights Differing Approaches Needed for Oomycete and Fungal Effectors.” <em>Front Plant Sci</em>, no. 6: 1168.</p>
</div>
<div id="ref-Petersen2011">
<p>Thomas Nordahl Petersen, Gunnar von Heijne, Soren Brunak, and Henrik Nielsen. 2011. “SignalP 4.0: Discriminating Signal Peptides from Transmembrane Regions.” <em>Nature Methods</em> 8 (10): 785–86.</p>
</div>
</div>
</div>
</div>
<div class="footnotes">
<hr />
<ol>
<li id="fn1"><p>A footnote here.<a href="#fnref1">↩</a></p></li>
</ol>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
