<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>SecretSanta: flexible pipelines for automated secretome prediction • SecretSanta</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">SecretSanta</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/SecretSanta-vignette.html">Vignette</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>SecretSanta: flexible pipelines for automated secretome prediction</h1>
                        <h4 class="author">Anna Gogleva</h4>
            
            <h4 class="date">2017-10-10</h4>
          </div>

    
    
<div class="contents">
<div id="background" class="section level2">
<h2 class="hasAnchor">
<a href="#background" class="anchor"></a>1. Background</h2>
<p>The <strong>SecretSanta</strong> package provides an R interface for the integrative prediction of extracellular proteins that are secreted via classical pathways.</p>
<p>Secretome prediction often involves multiple steps. Typically, it starts with prediction of short signal peptides at the N-terminal end of a protein (<strong>Figure 1 a</strong>). Next, it is crucial to ensure the absence of motifs and domains preventing the protein from being secreted despite the presence of the signal peptide. These sequences include transmembrane domains, short ER lumen retention signals and mitochondria/plastid targeting signals (<strong>Figure 1 b-d</strong>). The ultimate aim of a secretome prediction pipeline is to identify secreted proteins as shown in <strong>Figure 1 a</strong> and filer out those shown in <strong>Figure 1 b-d</strong>.</p>
<p></p>
<div class="figure" style="text-align: center">
<img src="motifs_and_domains.png" alt="\label{fig:motifs} **Figure 1.** Characteristic motifs, domains and 
their arrangements, helping to distinguish extracellular proteins from 
proteins retained inside the cell." width="400px"><p class="caption">
 <strong>Figure 1.</strong> Characteristic motifs, domains and their arrangements, helping to distinguish extracellular proteins from proteins retained inside the cell.
</p>
</div>
<p></p>
<p>A number of excellent command line tools and web-interfaces are available to perform predictions of individual motifs and domains (<a href="http://www.cbs.dtu.dk/services/SignalP/">signalp</a>, <a href="http://www.cbs.dtu.dk/services/TargetP/">targetp</a>, <a href="http://www.cbs.dtu.dk/services/TMHMM/">TMHMM</a>, <a href="https://github.com/fmaguire/WoLFPSort">WolfPsort</a>), however the interface allowing to combine the outputs in a single flexible workflow is lacking.</p>
<p></p>
<p>The <strong>SecretSanta</strong> package attempts to bridge this gap. It provides wrapper functions around existing command line tools for predictions of signal peptides and protein subcellular localization. The wrappers are designed to work together by producing standardized output as an instance of <code>CBSResult</code> superclass.</p>
<p></p>
<p>Objects of <code>CBSResult</code> class contain an <code>in_fasta</code> slot with the initial submitted sequences and an <code>out_fasta</code> slot with positive candidates after the method application. Each particular method then complements this simple structure with relevant slots. For example, outputs of the <code>signalp</code> function are organised in <code>SignalpResult</code> objects. Apart from <code>in_fasta</code> and <code>out_fasta</code> slots inherited from <code>CBSResult</code> class, <code>SignalpResult</code> objects contain three additional slots, relevant for the <strong>signalp</strong> method:</p>
<ul>
<li>
<code>sp_tibble</code> - parsed <strong>signalp</strong> tabular output for positive candidates;</li>
<li>
<code>mature_fasta</code> - mature sequence for the candidate secreted proteins, i.e sequences with cleaved N-terminal signal peptides;</li>
<li>
<code>sp_version</code> - version of <strong>signalp</strong> used to generate this object.</li>
</ul>
<p>The detailed organisation of class structure is shown in the <strong>Figure 2</strong>.</p>
<div class="figure" style="text-align: center">
<img src="class_structure.png" alt="\label{fig:motifs} **Figure 2.** SecretSanta class structure." width="600px"><p class="caption">
 <strong>Figure 2.</strong> SecretSanta class structure.
</p>
</div>
<p></p>
<p>This allows the user to easily pipe results between individual predictors to create flexible custom pipelines and also to compare predictions between similar methods. For instance, <strong>targetp</strong> and <strong>WoLFPsort</strong> for subcellular localization and multiple versions of <strong>signalp</strong> for signal peptide prediction.</p>
<p></p>
<p>To speed-up processing of large input fasta files initial steps of the pipeline are automatically run as an embarrassingly parallel process when the number of input sequences exceeds a certain limit.</p>
<p></p>
<p>Taken together <strong>SecretSanta</strong> provides a platform to build automated multi-step secretome prediction pipelines that can be applied to large protein sets to facilitate comparisons of secretomes across multiple species or under various conditions. For example, an earlier draft of this package was successfully used to compare the secreted protein complement of a plant pathogen throughout an infection time-course in plant roots <span class="citation">(Evangelisti E 2017)</span>.</p>
<p></p>
<p>To avoid confusion, the names of external command line tools will be shown in <strong>bold</strong>, and the corresponding R functions from the <strong>SecretSanta</strong> package will have standard <code>code highlighting</code>. For example:</p>
<ul>
<li>
<strong>signalp</strong> - CBS signalp and its associated derivative files;</li>
<li>
<code>signalp</code> - <strong>SecretSanta</strong> wrap function around <strong>signalp</strong>;</li>
</ul>
<p>Majority of the <strong>SecretSanta</strong> functions could be run in 2 modes:</p>
<ul>
<li>
<code>starter</code> - when initiating the secretome pipeline with the above function;</li>
<li>
<code>piper</code> - for downstream/intermediate steps, so that the function expects an output from another wrapper function.</li>
</ul>
</div>
<div id="installation-of-external-dependencies" class="section level2">
<h2 class="hasAnchor">
<a href="#installation-of-external-dependencies" class="anchor"></a>2. Installation of external dependencies</h2>
<p><strong>SecretSanta</strong> relies on a set of existing command line tools to predict secretd proteins. Please install them and configure according to the listed instructions. Due to limitations imposed by the external dependencies, some of <strong>SecretSanta</strong> wrapper functions will nott work on Windows, however are fully fuctional on Linux (currently testing on Mac).</p>
<p></p>
<div id="download-and-configure-external-dependencies" class="section level3">
<h3 class="hasAnchor">
<a href="#download-and-configure-external-dependencies" class="anchor"></a>2.1 Download and configure external dependencies</h3>
<p>to do: add automatic istallation of deps</p>
<div id="tools-for-prediction-of-signal-peptides-and-cleavage-sites" class="section level4">
<h4 class="hasAnchor">
<a href="#tools-for-prediction-of-signal-peptides-and-cleavage-sites" class="anchor"></a>Tools for prediction of signal peptides and cleavage sites:</h4>
<ul>
<li>
<strong>signalp-2.0</strong>
<ul>
<li>This version can run under IRIX, IRIX64, Linux, OSF1, SunOS.</li>
<li>Download stand alone signalp2.0 <a href="http://www.cbs.dtu.dk/cgi-bin/sw_request?signalp+2.0" class="uri">http://www.cbs.dtu.dk/cgi-bin/sw_request?signalp+2.0</a>
</li>
<li>Unpack the archive</li>
</ul>
<div class="sourceCode"><pre class="sourceCode sh"><code class="sourceCode bash"><span class="kw">tar</span> -zxvf signalp-2.0.Linux.tar.Z
<span class="kw">cd</span> signalp-2.0</code></pre></div>
<ul>
<li>Edit “General settings” at the top of the <strong>signalp</strong> file. Set value of ‘SIGNALP’ variable to be path to your <strong>signalp-2.0</strong> directory. Other variables usually do not require changes. We will not use plotting functions from signalp, so gnuplot, ppmtogif and ghostview are not required. For more details please check <code>signalp-2.0.readme</code>.</li>
<li>Since we want to be able to run different versions of <strong>signalp</strong>, including the legacy versions, it is important to be able to discriminate between them. R is oblivious to shell aliases, so we will simply rename the <strong>siganlp</strong> script:</li>
</ul>
<div class="sourceCode"><pre class="sourceCode sh"><code class="sourceCode bash"><span class="kw">mv</span> signalp siganlp2</code></pre></div>
</li>
</ul>
<p></p>
<ul>
<li>
<strong>signalp-3.0</strong>
<ul>
<li>This version will run on the most common UNIX platforms.</li>
<li>Download stand alone signalp3.0 <a href="http://www.cbs.dtu.dk/cgi-bin/sw_request?signalp+3.0" class="uri">http://www.cbs.dtu.dk/cgi-bin/sw_request?signalp+3.0</a>
</li>
<li>Unpack the archive</li>
</ul>
<div class="sourceCode"><pre class="sourceCode sh"><code class="sourceCode bash"><span class="kw">tar</span> -zxvf signalp-3.0.Linux.tar.Z
<span class="kw">cd</span> signalp-3.0</code></pre></div>
<ul>
<li>Similar to <strong>signalp-2.0</strong>, edit “General settings” at the top of the signalp file. Set value of ‘SIGNALP’ variable to be path to your <strong>signalp-3.0</strong> directory. Other variables usually do not require changes. For more details please check <code>signalp-3.0.readme</code>.</li>
<li>Rename <strong>signalp</strong> script to avoid further confusion between the versions:</li>
</ul>
<div class="sourceCode"><pre class="sourceCode sh"><code class="sourceCode bash"><span class="kw">mv</span> signalp signalp3</code></pre></div>
</li>
</ul>
<p></p>
<ul>
<li>
<strong>signalp-4.1</strong> - the most recent version
<ul>
<li>This version can run under Windows, OS X (Macintosh) and Linux.</li>
<li>Download stand alone signalp4.0 <a href="http://www.cbs.dtu.dk/cgi-bin/nph-sw_request?signalp" class="uri">http://www.cbs.dtu.dk/cgi-bin/nph-sw_request?signalp</a>
</li>
<li>Unpack the archive</li>
</ul>
<div class="sourceCode"><pre class="sourceCode sh"><code class="sourceCode bash"><span class="kw">tar</span> -zxvf signalp-4.1.Linux.tar.Z
<span class="kw">cd</span> signalp-4.1</code></pre></div>
<ul>
<li>Edit “General settings” at the top of the <strong>signalp</strong> file. Set values for ‘SIGNALP’ and ‘outputDir’ variables. For more details please check <code>signalp-4.1.readme</code>.</li>
<li>Rename <strong>signalp</strong> script to avoid further confusion between the versions:</li>
</ul>
<div class="sourceCode"><pre class="sourceCode sh"><code class="sourceCode bash"><span class="kw">mv</span> signalp signalp4</code></pre></div>
</li>
</ul>
<p></p>
</div>
<div id="tools-for-prediction-of-protein-subcellular-localization" class="section level4">
<h4 class="hasAnchor">
<a href="#tools-for-prediction-of-protein-subcellular-localization" class="anchor"></a>Tools for prediction of protein subcellular localization:</h4>
<ul>
<li>
<strong>taretp-1.1</strong>
<ul>
<li>
<strong>tatgetp-1.1</strong> will run on the most common UNIX platforms<br>
</li>
<li>Download stand alone targetp <a href="http://www.cbs.dtu.dk/cgi-bin/nph-sw_request?targetp" class="uri">http://www.cbs.dtu.dk/cgi-bin/nph-sw_request?targetp</a>
</li>
<li>Unpack the archive:</li>
</ul>
<div class="sourceCode"><pre class="sourceCode sh"><code class="sourceCode bash"><span class="kw">tar</span> -zxvf targetp-1.1b.Linux.tar.Z 
<span class="kw">cd</span> targetp-1.1</code></pre></div>
<ul>
<li>Edit the paragraph labeled “GENERAL SETTINGS, customize” at the top of the <strong>targetp</strong> file. Set values for ‘TARGETP’ and ‘TMP’ variables. Ensure, that the path to <strong>targetp</strong> does not exceed 60 characters, otherwise <strong>targetp-1.1</strong> might fail.</li>
</ul>
</li>
</ul>
<p></p>
<ul>
<li>
<strong>WoLFPsort</strong>
<ul>
<li>Clone WoLFPsort</li>
</ul>
<div class="sourceCode"><pre class="sourceCode sh"><code class="sourceCode bash"><span class="kw">git</span> clone https://github.com/fmaguire/WoLFPSort.git
<span class="kw">cd</span> WoLFPSort</code></pre></div>
<ul>
<li>Copy the binaries from the appropriate platform specific binary directory <code>./bin/binByPlatform/binary-?</code> to <code>./bin/</code>
</li>
<li>For more details please check the <code>INSTALL</code> file.</li>
<li>The most important script we need <strong>runWolfPsortSummary</strong> has a bulky name, we will rename it to <strong>wolfpsort</strong> for future convenience:</li>
</ul>
<div class="sourceCode"><pre class="sourceCode sh"><code class="sourceCode bash"><span class="kw">mv</span> runWolfPsortSummary wolfpsort</code></pre></div>
</li>
</ul>
<p></p>
</div>
<div id="tools-for-predicting-transmembrane-domains" class="section level4">
<h4 class="hasAnchor">
<a href="#tools-for-predicting-transmembrane-domains" class="anchor"></a>Tools for predicting transmembrane domains</h4>
<ul>
<li>
<strong>tmhmm-2.0</strong>
<ul>
<li>
<strong>tmhmm-2.0</strong> will run on the most common UNIX platforms<br>
</li>
<li>Download stand alone tmhmm (<a href="http://www.cbs.dtu.dk/cgi-bin/nph-sw_request?tmhmm" class="uri">http://www.cbs.dtu.dk/cgi-bin/nph-sw_request?tmhmm</a>)</li>
<li>Unpack the archive:</li>
</ul>
<div class="sourceCode"><pre class="sourceCode sh"><code class="sourceCode bash"><span class="kw">tar</span> -zxvf tmhmm-2.0c.Linux.tar.gz
<span class="kw">cd</span> tmhmm-2.0c</code></pre></div>
<ul>
<li>Set correct path for Perl 5.x in the first line of <code>bin/tmhmm</code> and <code>bin/tmhmmformat.pl</code> scripts.</li>
<li>For more details please check the <code>README</code> file.</li>
</ul>
</li>
</ul>
<p></p>
</div>
</div>
<div id="organise-access-to-the-external-dependencies" class="section level3">
<h3 class="hasAnchor">
<a href="#organise-access-to-the-external-dependencies" class="anchor"></a>2.2 Organise access to the external dependencies</h3>
<p>Two options are possible:</p>
<div id="option-a-all-the-external-dependencies-are-accessible-from-any-location-" class="section level4">
<h4 class="hasAnchor">
<a href="#option-a-all-the-external-dependencies-are-accessible-from-any-location-" class="anchor"></a>Option A: all the external dependencies are accessible from any location.</h4>
<p>This requires modification of <code>$PATH</code> environment variable. To make the change permanent, edit <code>.profile</code>:</p>
<p>Open ./profile:</p>
<div class="sourceCode"><pre class="sourceCode sh"><code class="sourceCode bash"><span class="kw">gedit</span> ~/.profile</code></pre></div>
<p>Add a line with all the path exports. In this example all the dependencies are installed in the <code>my_tool</code> directory:</p>
<div class="sourceCode"><pre class="sourceCode sh"><code class="sourceCode bash"><span class="kw">export</span> <span class="ot">PATH=</span>
<span class="st">"/home/my_tools/signalp-4.1:\</span>
<span class="st">/home/my_tools/signalp-2.0:\</span>
<span class="st">/home/my_tools/signalp-3.0:\</span>
<span class="st">/home/my_tools/targetp-1.1:\</span>
<span class="st">/home/tmhmm-2.0c/bin:\</span>
<span class="st">/home/my_tools/WoLFPSort/bin:\</span>
<span class="ot">$PATH</span><span class="st">"</span></code></pre></div>
<p>Reload <code>/.profile</code>:</p>
<div class="sourceCode"><pre class="sourceCode sh"><code class="sourceCode bash"><span class="kw">.</span> <span class="kw">~/.profile</span></code></pre></div>
<p>Reboot to make changes visible to R. Alternatively - restart the R session.</p>
</div>
<div id="option-b-paths-to-external-dependencies-are-supplied-in-a-separate-file" class="section level4">
<h4 class="hasAnchor">
<a href="#option-b-paths-to-external-dependencies-are-supplied-in-a-separate-file" class="anchor"></a>Option B: paths to external dependencies are supplied in a separate file:</h4>
<p>This option should be used when <code>$PATH</code> variable can not be modified. Create a 2-column space-separated text file with listed paths for all the external dependencies.</p>
<table class="table">
<caption>
<strong>Table1</strong>: Sample paths to external dependencies</caption>
<thead><tr class="header">
<th align="left">tool</th>
<th align="left">path</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">signalp4</td>
<td align="left">/home/tools/signalp-4.1/signalp</td>
</tr>
<tr class="even">
<td align="left">signalp3</td>
<td align="left">/home/tools/signalp-3.0/signalp</td>
</tr>
<tr class="odd">
<td align="left">signalp2</td>
<td align="left">/home/tools/signalp-2.0/signalp</td>
</tr>
<tr class="even">
<td align="left">targetp</td>
<td align="left">/home/tools/targetp-1.1/targetp</td>
</tr>
<tr class="odd">
<td align="left">TMHMM</td>
<td align="left">/home/tools/tmhmm-2.0c/bin/tmhmm</td>
</tr>
<tr class="even">
<td align="left">WoLFPSORT</td>
<td align="left">/home/tools/WoLFPSort-master/bin/wolfpsort</td>
</tr>
</tbody>
</table>
<p>For external tool names please use simple character strings without version numbers. The only exception here are different versions of <strong>signalp</strong>, which require version number in one-digit format in the <code>tool</code> column to be distinguishable.</p>
</div>
</div>
<div id="check-that-all-the-required-dependencies-are-installed-correctly" class="section level3">
<h3 class="hasAnchor">
<a href="#check-that-all-the-required-dependencies-are-installed-correctly" class="anchor"></a>2.3 Check that all the required dependencies are installed correctly</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(SecretSanta)</code></pre></div>
<p>The <code>manage_paths</code> function will run small test jobs to check that all the external dependencies are functional.</p>
<p>In case all the dependencies are available globally (see <strong>Option A</strong>):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">check1 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/manage_paths.html">manage_paths</a></span>(<span class="dt">in_path =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p>It is also possible to run just a single test for a specific dependency in question:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">check1 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/manage_paths.html">manage_paths</a></span>(<span class="dt">in_path =</span> <span class="ot">TRUE</span>, <span class="dt">test_mode =</span> <span class="st">'signalp2'</span>)
## checking dependencies accessible via $PATH
## signalp2 run completed</code></pre></div>
<p>In case we have paths organised in a separate file (see <strong>Option B</strong>):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">check2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/manage_paths.html">manage_paths</a></span>(
    <span class="dt">in_path =</span> <span class="ot">FALSE</span>,
    <span class="dt">path_file =</span> <span class="kw">system.file</span>(<span class="st">"extdata"</span>,
    <span class="st">"sample_paths"</span>,
    <span class="dt">package =</span> <span class="st">"SecretSanta"</span>)
    )</code></pre></div>
<p>Note: <code>manage_paths</code> is case insensitive and will convert all the tool names to the lower case.</p>
</div>
</div>
<div id="individual-methods" class="section level2">
<h2 class="hasAnchor">
<a href="#individual-methods" class="anchor"></a>3. Individual methods</h2>
<div id="signalp" class="section level3">
<h3 class="hasAnchor">
<a href="#signalp" class="anchor"></a>3.1 Signalp</h3>
<p><strong>signalp</strong> is a software tool to predict classical signal peptides and cleavage sites <span class="citation">(Henrik Nielsen and Heijne 1997)</span> in eukaryotes and bacteria. The method combines prediction of signal peptides and cleavage cites based on a combination of artificial neural networks and hidden Markov models. The latter can distinguish between signal peptides and non-cleaved signal anchors <span class="citation">(Henrik Nielsen and Heijne 1997)</span>.</p>
<p>Currently <code>signalp</code> function from SecretSanta provides an interface for the three recent versions of <strong>signalp</strong>. The reason behind this is that <strong>signalp-4</strong> and <strong>signalp-4.1</strong> are not sensitive enough to predict certain classes of secreted oomycete and fungal effectors <span class="citation">(Sperschneider J 2015)</span>. Thus, providing access to the older versions allows greater flexibility for building versatile prediction pipelines for a wider taxonomic range of species.</p>
<ul>
<li>
<strong>signalp-2.0</strong> <span class="citation">(Nielsen and Krogh 1998)</span>
</li>
<li>
<strong>signalp-3.0</strong> <span class="citation">(Dyrl�v Bendtsen and Brunak 2004)</span>
</li>
<li>
<strong>signalp-4.0</strong> <span class="citation">(Thomas Nordahl Petersen and Nielsen. 2011)</span>
</li>
</ul>
<p><code>signalp</code> requires specifying <code>organism</code> argument; possible values include:</p>
<ul>
<li>
<em>‘euk’</em> - for eukaryotes;</li>
<li>
<em>‘gram+’</em> - for gram-positive bacteria;</li>
<li>
<em>‘gram-’</em> - for gram-negative bacteria.</li>
</ul>
<p>To run <code>signalp</code> prediction, first read fasta file with amino acid sequences and store its contents in a separate variable:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">aa &lt;-<span class="st"> </span><span class="kw">readAAStringSet</span>(<span class="kw">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"sample_prot_100.fasta"</span>,
                                    <span class="dt">package =</span> <span class="st">"SecretSanta"</span>))</code></pre></div>
<p>Initialize object of CBSResult class with <code>aa</code> as <code>in_fasta</code> attribute.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">inp &lt;-<span class="st"> </span><span class="kw"><a href="../reference/CBS_methods.html">CBSResult</a></span>(<span class="dt">in_fasta =</span> aa)
inp</code></pre></div>
<pre><code>## An object of class "CBSResult"
## Slot "in_fasta":
##   A AAStringSet instance of length 100
##       width seq                                        names               
##   [1]    94 MVEPSSPECVVQIIKDEKLK...LLVCGFIYFAVLQSWLLES ALI_PLTG_1 Hypoth...
##   [2]   240 MTRQGYLIVHDKRSRPTVRY...VTPHYVPTHFGRGNNQEAH ALI_PLTG_2 Unknow...
##   [3]   533 MQPTSSSAPQSDVVVADPPS...QQQQEQQANSTALKNERSI ALI_PLTG_3 Hypoth...
##   [4]   456 QFLSELRKVSVSTKLDGVQR...QSQQNSRPREGCCSSCIIF ALI_PLTG_4 Hypoth...
##   [5]   210 MKLSSVCTIIFGLVFIDFNN...DGSVTTAAQKAAALVKANA ALI_PLTG_5 Cutin ...
##   ...   ... ...
##  [96]   650 MQRLVQRDKDAPVVREGDDD...DRLSLNALLEAFPPMNSRL ALI_PLTG_96 Prote...
##  [97]   236 MSTKRSRSAPILRVKKLTPE...DDDNSAKVLEWINAFIESN ALI_PLTG_97 Deoxy...
##  [98]   239 MSLNYLAHTNMRRARENFLK...KTYIEYEAVENNTLNKTEG ALI_PLTG_98 Hypot...
##  [99]   566 MCTSFSDQEDKKLVVLATEY...PRLDRKFFGLQVAPPPTSQ ALI_PLTG_99 Histo...
## [100]   180 MDNYYTSVQLLLDLQLKGLC...DPHRTFLTDLVGVELQVMG ALI_PLTG_100 Tran...
## 
## Slot "out_fasta":
##   A AAStringSet instance of length 0</code></pre>
<p>Since this is the first step in our secretome prediction pipeline, we will run <code>signalp</code> in a <em>starter</em> mode. Here we select version number 2 (<strong>signalp-2.0</strong>) and assume that signalp is accessible globally, so no path files should be specified. If <strong>signalp</strong> can not be accessed globally, this will result in an error message.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">step1_sp2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/signalp.html">signalp</a></span>(inp, <span class="dt">version =</span> <span class="dv">2</span>,
                    <span class="dt">organism =</span> <span class="st">'euk'</span>,
                    <span class="dt">run_mode =</span> <span class="st">"starter"</span>)
## Version used... signalp2
## running signalp locally...
## 2 sequences to be truncated
## Ok for single processing
## Submitted sequences... 100
## signalp &lt; 4, calling parser for the output...
## Candidate sequences with signal peptides... 8</code></pre></div>
<p>The above code under the hood runs <strong>signalp-2.0</strong> prediction and outputs result as an instance of <code>SignalpResult</code> class, which belongs to a more generic <code><a href="../reference/CBS_methods.html">CBSResult()</a></code> superclass having just 2 slots: <code>in_fasta</code> and <code>out_fasta</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">class</span>(step1_sp2)</code></pre></div>
<pre><code>## [1] "SignalpResult"
## attr(,"package")
## [1] "SecretSanta"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">slotNames</span>(step1_sp2)</code></pre></div>
<pre><code>## [1] "mature_fasta" "sp_version"   "sp_tibble"    "in_fasta"    
## [5] "out_fasta"</code></pre>
<p>The <code>SignalpResult</code> object contains 5 slots:</p>
<ul>
<li><p><code>in_fasta</code> - original set of input amino acid sequences;</p></li>
<li><p><code>sp_tibble</code> - parsed <strong>signalp</strong> tabular output for positive candidates;</p></li>
<li><p><code>out_fasta</code> - full length amino acid sequences of positive candidates;</p></li>
<li><p><code>mature_fasta</code> - mature sequence for the candidate secreted proteins, i.e sequences with cleaved N-terminal signal peptides;</p></li>
<li><p><code>sp_version</code> - version of <strong>signalp</strong> used to generate this object.</p></li>
</ul>
<p>You can use accessor methods to get contents of individual slots:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/CBS_methods.html">getInfasta</a></span>(step1_sp2)</code></pre></div>
<pre><code>##   A AAStringSet instance of length 100
##       width seq                                        names               
##   [1]   174 MDALRESEDLRIHSPTESSS...LRTRIEMSFGLLANNGRFI ALI_PLTG_30
##   [2]   456 MLASKIVPLRRAASTVGARN...ALGSDAKYAGEHFRNPSKM ALI_PLTG_88
##   [3]   154 MYLKFAFMATVATAVVNIAI...AKLLLAYLNDFKRTSGLRS ALI_PLTG_73
##   [4]   566 MCTSFSDQEDKKLVVLATEY...PRLDRKFFGLQVAPPPTSQ ALI_PLTG_99
##   [5]   495 MVVRVLLVVLALLAVGVQSK...AIVTIGASDGRVRYMNPPS ALI_PLTG_38
##   ...   ... ...
##  [96]   446 MVSPQPQDDACCVKTPLTKS...QLETHLQQRLAVLARSVSV ALI_PLTG_13
##  [97]   533 MQPTSSSAPQSDVVVADPPS...QQQQEQQANSTALKNERSI ALI_PLTG_3
##  [98]   251 MVLLPFWSRAKRLRQWLLGA...ETCVIAVYDPSIRSVHNIC ALI_PLTG_84
##  [99]    90 MSSAQYVSTKVSVDMFNGGN...HMDADYHHVVDNCEEAWVA ALI_PLTG_48
## [100]    64 RRIWTDVATAFVSSCREFCS...TGATYRQTCWHQCRTGFSC ALI_PLTG_29</code></pre>
<p>Please note that <code>signlap</code> truncates all the input sequences longer than 2000 a.a, by default. In this case ’_truncated’ is added to the original sequence ids to keep track of the changes. An alternative option is to discard long sequences completely before the analysis, to do so set <code>truncate = FALSE</code> when running <code>signalp</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/SignalpResult_methods.html">getSPtibble</a></span>(step1_sp2)</code></pre></div>
<pre><code>## # A tibble: 8 x 9
##       gene_id  Cmax  Cpos  Ymax  Ypos  Smax  Spos Smean     Prediction
##        &lt;fctr&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;         &lt;fctr&gt;
## 1 ALI_PLTG_73 0.489    22 0.316    22 0.927     3 0.721 Signal peptide
## 2 ALI_PLTG_38 0.572    20 0.696    20 0.994     9 0.954 Signal peptide
## 3 ALI_PLTG_83 0.393    34 0.506    20 0.956    13 0.846 Signal peptide
## 4 ALI_PLTG_23 0.641    22 0.683    22 0.969     8 0.803 Signal peptide
## 5 ALI_PLTG_81 0.305    33 0.388    33 0.826    25 0.471 Signal peptide
## 6 ALI_PLTG_72 0.887    22 0.804    22 0.986     8 0.805 Signal peptide
## 7  ALI_PLTG_5 1.000    25 0.687    25 0.906     3 0.637 Signal peptide
## 8 ALI_PLTG_59 0.210    17 0.288    19 0.948     1 0.666 Signal peptide</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/CBS_methods.html">getOutfasta</a></span>(step1_sp2)</code></pre></div>
<pre><code>##   A AAStringSet instance of length 8
##     width seq                                          names               
## [1]   154 MYLKFAFMATVATAVVNIAIG...YAKLLLAYLNDFKRTSGLRS ALI_PLTG_73
## [2]   495 MVVRVLLVVLALLAVGVQSKA...FAIVTIGASDGRVRYMNPPS ALI_PLTG_38
## [3]   164 MVSTSRVFALLLLPSPSARIF...SMVEHIKTTKRVIDEVQDHV ALI_PLTG_83
## [4]   653 MKIVALVTFCIATLDSSIVFA...SEDPVYPLVKEYSDVVSKHP ALI_PLTG_23
## [5]   133 MNFLKSKKTSTQATLLVSPLI...MGGLMPNLHVDLENGIGNYP ALI_PLTG_81
## [6]   118 MRLKFSILIFGAVLLATTTNA...NIRYASMLQDFLNTYHRRGV ALI_PLTG_72
## [7]   210 MKLSSVCTIIFGLVFIDFNNV...SDGSVTTAAQKAAALVKANA ALI_PLTG_5
## [8]   125 MLMSTPAALMLRLAQSSLKPT...DQGAHVATSGTSNNVLLWHK ALI_PLTG_59</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/SignalpResult_methods.html">getMatfasta</a></span>(step1_sp2)</code></pre></div>
<pre><code>##   A AAStringSet instance of length 8
##     width seq                                          names               
## [1]   133 NVIDLPMPLTLESPNFVESAA...YAKLLLAYLNDFKRTSGLRS ALI_PLTG_73
## [2]   476 KATRVRKSWTAYTSDEKEIYL...FAIVTIGASDGRVRYMNPPS ALI_PLTG_38
## [3]   131 ITTKLKQFNGTGFPAWGVQDK...SMVEHIKTTKRVIDEVQDHV ALI_PLTG_83
## [4]   632 AECTVDELTEISTIYSEAMTD...SEDPVYPLVKEYSDVVSKHP ALI_PLTG_23
## [5]   101 YGPPADVRPMVVRMKSDYRSY...MGGLMPNLHVDLENGIGNYP ALI_PLTG_81
## [6]    97 DDNSNKRLLRRQEKTDTTNGD...NIRYASMLQDFLNTYHRRGV ALI_PLTG_72
## [7]   186 QCSDVHVVFARGSGEAAGLGI...SDGSVTTAAQKAAALVKANA ALI_PLTG_5
## [8]   109 SLKPTVGKAITVGGNHQIPTK...DQGAHVATSGTSNNVLLWHK ALI_PLTG_59</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/SignalpResult_methods.html">getSPversion</a></span>(step1_sp2)</code></pre></div>
<pre><code>## [1] 2</code></pre>
<p>Next, imagine we would like to run all the <strong>signalp</strong> version on the same input for comparison. We can do this by simply changing <code>version</code> value:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">step1_sp3 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/signalp.html">signalp</a></span>(inp, 
                    <span class="dt">version =</span> <span class="dv">3</span>,
                    <span class="dt">organism =</span> <span class="st">'euk'</span>,
                    <span class="dt">run_mode =</span> <span class="st">"starter"</span>)
## Version used... signalp3
## running signalp locally...
## 2 sequences to be truncated
## Ok for single processing
## Submitted sequences... 100
## signalp &lt; 4, calling parser for the output...
## Candidate sequences with signal peptides... 8
step1_sp4 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/signalp.html">signalp</a></span>(inp,
                    <span class="dt">version =</span> <span class="dv">4</span>,
                    <span class="dt">organism =</span> <span class="st">'euk'</span>,
                    <span class="dt">run_mode =</span> <span class="st">"starter"</span>)
## Version used... signalp4
## running signalp locally...
## 2 sequences to be truncated
## Ok for single processing
## Submitted sequences... 100
## Candidate sequences with signal peptides... 5</code></pre></div>
<p>In this case <code>signalp-4.1</code> version resulted in fewer candidates. Please note, that despite differences in the output format generated by multiple versions of <strong>signalp</strong>, <code>signalp</code> returns <code>sp_tibble</code> in a standard format.</p>
<p>In order to pipe result from different versions of <strong>signalp</strong> just switch to the <code>run_mode = 'piper'</code> for the second and other downstream steps. In this case <code>signalp</code> will run the analysis using contents of <code>out_fasta</code> slot as an input. Say, we want to do the following piping: <strong>signalp2</strong> -&gt; <strong>signalp3</strong> -&gt; <strong>signalp4</strong>.</p>
<p></p>
<p>We will re-use the <code>step1_sp2</code> object generated earlier in the <code>starter</code> mode for the <strong>signalp2</strong> -&gt; <strong>signalp3</strong> piping operation:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">step2_sp3 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/signalp.html">signalp</a></span>(step1_sp2,
                    <span class="dt">version =</span> <span class="dv">3</span>,
                    <span class="dt">organism =</span> <span class="st">'euk'</span>,
                    <span class="dt">run_mode =</span> <span class="st">"piper"</span>)
## Version used... signalp3
## running signalp locally...
## Ok for single processing
## Submitted sequences... 8
## signalp &lt; 4, calling parser for the output...
## Candidate sequences with signal peptides... 8</code></pre></div>
<p>Similar with the <strong>signalp3</strong> -&gt; <strong>signalp4</strong> piping:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">step3_sp4 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/signalp.html">signalp</a></span>(step2_sp3,
                    <span class="dt">version =</span> <span class="dv">4</span>,
                    <span class="dt">organism =</span> <span class="st">'euk'</span>,
                    <span class="dt">run_mode =</span> <span class="st">"piper"</span>)
## Version used... signalp4
## running signalp locally...
## Ok for single processing
## Submitted sequences... 8
## Candidate sequences with signal peptides... 5</code></pre></div>
<p>With the input fasta files containing more than 1000 sequences, <code>signalp</code> will automatically switch to parallel mode. It will split the input into smaller chunks and run prediction as an embarrassingly parallel process using specified number of CPUs (see <code>cores</code> argument). Execution time depends on the number of CPUs available and the input file size. With 48 CPUs it takes ~2 minutes to run <code>signalp</code> on the input fasta file with more than 40’000 sequences.</p>
</div>
<div id="tmhmm" class="section level3">
<h3 class="hasAnchor">
<a href="#tmhmm" class="anchor"></a>3.2 TMHMM</h3>
<p><strong>TMHMM</strong> predicts transmembrane a-helices and identifies integral membrane proteins based on HMMs [<span class="citation">Krogh, Bjorn Larsson, and Sonnhammer (2001)</span>; Sonnhammer1998]. It is important to exclude proteins with transmembrane domains located after signal peptide, as they will be retained in the membrane. Since <strong>TMHMM</strong> is often unable to distinguish between N-terminal signal peptides and transmembrane domains, it is recommended to run <code>tmhmmm</code> on the mature sequences obtained after running <code>signalp</code> function.</p>
<p><code>tmhmm</code> function can handle input objects of <code>SignalpResult</code> class with non-empty <code>mature_fasta</code> slot.</p>
<p>Here we will use output of the <code>signalp</code> function as an input for <code>tmhmm</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tm &lt;-<span class="st"> </span><span class="kw"><a href="../reference/tmhmm.html">tmhmm</a></span>(step1_sp2, <span class="dt">TM =</span> <span class="dv">0</span>)
## running TMHMM locally...
## Submitted sequences... 8
## Candidates with signal peptides and 0 TM domains in mature seq 8</code></pre></div>
<p>Attempts to run <code>tmhmm</code> on the <code>CBSResult</code> object lacking <code>mature_fasta</code> slot will produce an error:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tm2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/tmhmm.html">tmhmm</a></span>(inp, <span class="dt">TM =</span> <span class="dv">0</span>)
## Error in tmhmm(inp, TM = 0): input_object does not belong to SignalpResult class</code></pre></div>
<p><code>tmhmm</code> outputs instances of <code>TMhmmResult</code> class, inheriting from <code>CBSResult</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">class</span>(tm)</code></pre></div>
<pre><code>## [1] "TMhmmResult"
## attr(,"package")
## [1] "SecretSanta"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">slotNames</span>(tm)</code></pre></div>
<pre><code>## [1] "in_mature_fasta"  "out_mature_fasta" "tm_tibble"       
## [4] "in_fasta"         "out_fasta"</code></pre>
<p><code>TMhmmResult</code> object contains 5 slots:</p>
<ul>
<li><p><code>in_fasta</code> and <code>out_fasta</code> are common attributes of all classes inheriting from the <code>CBSResult</code> class;</p></li>
<li><p><code>tm_tibble</code> - parsed <strong>tmhmm</strong> tabular output for candidates not having more transmembrane domains than specified by <code>TM</code> threshold;</p></li>
<li><p><code>in_mature_fasta</code> - mature sequences used as an input for <code>tmhmm</code></p></li>
<li><p><code>out_mature_fasta</code> - outputted mature sequences not having more transmembrane domains than specified by <code>TM</code> threshold.</p></li>
</ul>
<p>To get contents of individual slots you could use accessor functions:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/TMhmmResult_methods.html">getTMtibble</a></span>(tm)</code></pre></div>
<pre><code>## # A tibble: 8 x 6
##       gene_id length ExpAA First60 PredHel Topology
##         &lt;chr&gt;  &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;    &lt;chr&gt;
## 1 ALI_PLTG_73    133  0.00    0.00       0        o
## 2 ALI_PLTG_38    476  0.13    0.09       0        o
## 3 ALI_PLTG_83    131  1.44    0.02       0        o
## 4 ALI_PLTG_23    632  0.00    0.00       0        o
## 5 ALI_PLTG_81    101  0.00    0.00       0        o
## 6 ALI_PLTG_72     97  0.00    0.00       0        i
## 7  ALI_PLTG_5    186  1.29    1.16       0        o
## 8 ALI_PLTG_59    109  0.00    0.00       0        o</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/TMhmmResult_methods.html">getInMatfasta</a></span>(tm)</code></pre></div>
<pre><code>##   A AAStringSet instance of length 8
##     width seq                                          names               
## [1]   133 NVIDLPMPLTLESPNFVESAA...YAKLLLAYLNDFKRTSGLRS ALI_PLTG_73
## [2]   476 KATRVRKSWTAYTSDEKEIYL...FAIVTIGASDGRVRYMNPPS ALI_PLTG_38
## [3]   131 ITTKLKQFNGTGFPAWGVQDK...SMVEHIKTTKRVIDEVQDHV ALI_PLTG_83
## [4]   632 AECTVDELTEISTIYSEAMTD...SEDPVYPLVKEYSDVVSKHP ALI_PLTG_23
## [5]   101 YGPPADVRPMVVRMKSDYRSY...MGGLMPNLHVDLENGIGNYP ALI_PLTG_81
## [6]    97 DDNSNKRLLRRQEKTDTTNGD...NIRYASMLQDFLNTYHRRGV ALI_PLTG_72
## [7]   186 QCSDVHVVFARGSGEAAGLGI...SDGSVTTAAQKAAALVKANA ALI_PLTG_5
## [8]   109 SLKPTVGKAITVGGNHQIPTK...DQGAHVATSGTSNNVLLWHK ALI_PLTG_59</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/TMhmmResult_methods.html">getOutMatfasta</a></span>(tm)</code></pre></div>
<pre><code>##   A AAStringSet instance of length 8
##     width seq                                          names               
## [1]   133 NVIDLPMPLTLESPNFVESAA...YAKLLLAYLNDFKRTSGLRS ALI_PLTG_73
## [2]   476 KATRVRKSWTAYTSDEKEIYL...FAIVTIGASDGRVRYMNPPS ALI_PLTG_38
## [3]   131 ITTKLKQFNGTGFPAWGVQDK...SMVEHIKTTKRVIDEVQDHV ALI_PLTG_83
## [4]   632 AECTVDELTEISTIYSEAMTD...SEDPVYPLVKEYSDVVSKHP ALI_PLTG_23
## [5]   101 YGPPADVRPMVVRMKSDYRSY...MGGLMPNLHVDLENGIGNYP ALI_PLTG_81
## [6]    97 DDNSNKRLLRRQEKTDTTNGD...NIRYASMLQDFLNTYHRRGV ALI_PLTG_72
## [7]   186 QCSDVHVVFARGSGEAAGLGI...SDGSVTTAAQKAAALVKANA ALI_PLTG_5
## [8]   109 SLKPTVGKAITVGGNHQIPTK...DQGAHVATSGTSNNVLLWHK ALI_PLTG_59</code></pre>
</div>
<div id="targetp" class="section level3">
<h3 class="hasAnchor">
<a href="#targetp" class="anchor"></a>3.3 Targetp</h3>
<p><strong>Targetp</strong> predicts the subcellular localization of secreted eukaryotic proteins based on the presence of signal peptide (SP), chloroplast transit peptides (cTP) or mitochondrial targeting peptides (mTP) in the N-terminus <span class="citation">(Olof Emanuelsson and Heijne 2008)</span>. Including <strong>targetp</strong> in the pipeline can provide additional evidence that a protein with a predicted signal peptide is not targeted to plastids or mitochondria and is indeed extracellular. The <code>targetp</code> function accepts input objects belonging to the <code>CBSResult</code> class.</p>
<p><code>targetp</code> requires specifying <code>network_type</code> to run correctly. Possible values include:</p>
<ul>
<li>
<em>‘P’</em> - for plants;</li>
<li>
<em>‘N’</em> - for non_plants.</li>
</ul>
<p>It is possible to run <code>targetp</code> in both <em>piper</em> and <em>starter</em> modes. Imagine, we want to start our pipeline by running <code>targetp</code> using already created <code>inp</code> object as an input:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tp &lt;-<span class="st"> </span><span class="kw"><a href="../reference/targetp.html">targetp</a></span>(inp,
        <span class="dt">network =</span> <span class="st">'N'</span>,  <span class="co"># for non-plant networks</span>
        <span class="dt">run_mode =</span> <span class="st">'starter'</span>)
## running targetp locally...
## Ok for single processing
## Number of submitted sequences... 100
## Number of candidate secreted sequences 10</code></pre></div>
<p>Alternatively, we can run targetp on the output of other functions, for example <code>signalp</code> in the <em>piper</em> mode:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tp_pipe &lt;-<span class="st"> </span><span class="kw"><a href="../reference/targetp.html">targetp</a></span>(step1_sp2,
                    <span class="dt">network =</span> <span class="st">'N'</span>,
                    <span class="dt">run_mode =</span> <span class="st">'piper'</span>)
## running targetp locally...
## Ok for single processing
## Number of submitted sequences... 8
## Number of candidate secreted sequences 6</code></pre></div>
<p>In both cases <code>targetp</code> will output an object of <code>TargetpResult</code> class with the following slots:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">class</span>(tp)</code></pre></div>
<pre><code>## [1] "TargetpResult"
## attr(,"package")
## [1] "SecretSanta"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">slotNames</span>(tp)</code></pre></div>
<pre><code>## [1] "tp_tibble" "in_fasta"  "out_fasta"</code></pre>
<ul>
<li>
<code>in_fasta</code> and <code>out_fasta</code> are common attributes of all classes inheriting from <code>CBSResult</code> class;</li>
<li>
<code>tp_tibble</code> - parsed tragetp tabular output for candidates not targeted to mitochondria or plastids, i.e most likely to be secreted.</li>
</ul>
<p>To access the <code>tp_tibble</code> slot use the <code>getTPtibble</code> method:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/TargetpResult_methods.html">getTPtibble</a></span>(tp)</code></pre></div>
<pre><code>## # A tibble: 10 x 7
##        gene_id length   mTP    sp other TP_localization    RC
##         &lt;fctr&gt;  &lt;int&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;          &lt;fctr&gt; &lt;int&gt;
##  1  ALI_PLTG_5    210 0.027 0.944 0.096               S     1
##  2 ALI_PLTG_19    957 0.176 0.359 0.306               S     5
##  3 ALI_PLTG_23    653 0.025 0.953 0.072               S     1
##  4 ALI_PLTG_38    495 0.123 0.951 0.012               S     1
##  5 ALI_PLTG_69    164 0.138 0.527 0.267               S     4
##  6 ALI_PLTG_72    118 0.086 0.877 0.054               S     2
##  7 ALI_PLTG_73    154 0.035 0.938 0.108               S     1
##  8 ALI_PLTG_74    419 0.024 0.922 0.257               S     2
##  9 ALI_PLTG_77    347 0.090 0.475 0.365               S     5
## 10 ALI_PLTG_81    133 0.134 0.581 0.133               S     3</code></pre>
<p><code>targetp</code> will automatically switch to a parallel mode if the number of input sequences is greater than 1000.</p>
</div>
<div id="wolfpsort" class="section level3">
<h3 class="hasAnchor">
<a href="#wolfpsort" class="anchor"></a>3.4 WoLFPSORT</h3>
<p><strong>WoLFPSORT</strong> - predicts protein subcellular localization based on the PSORT principle. It converts amino acid sequences into numerical localization features based on sorting signals, amino acid composition and functional motifs. The converted data is then classified based on k-nearest neighbor algorithm <span class="citation">(Paul Horton and Nakai 2006)</span>.</p>
<p><code>wolfpsort</code> requires <code>organism</code> parameter. Possible values include:</p>
<ul>
<li>
<em>‘plant’</em>;</li>
<li>
<em>‘animal’</em>;</li>
<li>
<em>‘fungi’</em>.</li>
</ul>
<p>The <code>wolfpsort</code> function accepts objects of <code>CBSResult</code> class with non-empty <code>out_fasta</code> slot, so it is recommended to run this prediction on later stages of the pipeline.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">wlf &lt;-<span class="st"> </span><span class="kw"><a href="../reference/wolfpsort.html">wolfpsort</a></span>(step1_sp2, <span class="dt">organism =</span> <span class="st">'fungi'</span>)
## running WoLF PSORT locally...
## Number of submitted sequences... 8
## Candidate sequences with extracellular localisation... 5</code></pre></div>
<p><code>wolfpsort</code> returns an object with 3 slots:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">class</span>(wlf)</code></pre></div>
<pre><code>## [1] "WolfResult"
## attr(,"package")
## [1] "SecretSanta"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">slotNames</span>(wlf)</code></pre></div>
<pre><code>## [1] "wolf_tibble" "in_fasta"    "out_fasta"</code></pre>
<ul>
<li>
<code>in_fasta</code> and <code>out_fasta</code> are common attributes of all classes inheriting from <code>CBSResult</code> class;</li>
<li>
<code>wolf_tibble</code> - parsed WoLFPsort tabular output for sequences with ‘extracellular’ predicted to be the most probable subcellular localization.</li>
</ul>
<p>To access contents of the <code>wolf_tibble</code> slot use <code>getWOlFtibble</code> method:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/WolfResult_methods.html">getWOLFtibble</a></span>(wlf)</code></pre></div>
<pre><code>## # A tibble: 5 x 2
##       gene_id localization
##        &lt;fctr&gt;       &lt;fctr&gt;
## 1 ALI_PLTG_73         extr
## 2 ALI_PLTG_38         extr
## 3 ALI_PLTG_23         extr
## 4 ALI_PLTG_81         extr
## 5  ALI_PLTG_5         extr</code></pre>
<p>Since <code>wolfpsort</code> has the same purpose as <code>targetp</code> it might be useful to run these functions side by side tp compare the obtained results (see <a href="#Targetp">targetp results</a>).</p>
</div>
<div id="check-the-presence-of-terminal-er-retention-motifs" class="section level3">
<h3 class="hasAnchor">
<a href="#check-the-presence-of-terminal-er-retention-motifs" class="anchor"></a>3.5 Check the presence of terminal ER-retention motifs</h3>
<p>In addition to having signal peptides, some proteins have an ER-retention signal in the C-terminal domain that prevents protein from being secreted outside the cell. There are at least 2 known ER-retention motifs (<em>KDEL</em> and <em>HDEL</em>) <span class="citation">(Munro S 1987)</span>.</p>
<p>The <code>check_khdel</code> function uses simple pattern matching to scan amino acid sequences and remove those with an ER retention signal in the C-terminus. To run it, simply pass a <code>CBSResult</code> object as an input. You can run <code>check_khdel</code> in either <em>piper</em> or <em>starter</em> mode. The function does not rely on external dependencies, so providing path container is not required.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">er_result &lt;-<span class="st"> </span><span class="kw"><a href="../reference/check_khdel.html">check_khdel</a></span>(step1_sp2, <span class="dt">run_mode =</span> <span class="st">'piper'</span>)
## checking for terminal ER retention signals...
## Submitted sequences... 8
## Sequences with terminal ER retention signals detected... 0
## Candidate without terminal ER retention signals detected... 8</code></pre></div>
<p>This will return and object with 3 slots:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">slotNames</span>(er_result)</code></pre></div>
<pre><code>## [1] "retained"  "in_fasta"  "out_fasta"</code></pre>
<ul>
<li>
<code>in_fasta</code> and <code>out_fasta</code> are common attributes of all classes inheriting from <code>CBSResult</code> class;</li>
<li>
<code>retained</code> - all the sequences with C-terminal ‘(K/H)DEL’ motifs, in case you would like to have a look at them.</li>
</ul>
</div>
<div id="m-slicer" class="section level3">
<h3 class="hasAnchor">
<a href="#m-slicer" class="anchor"></a>3.6 M-slicer</h3>
<p>This is an experimental option. The <code>m_slicer</code> function takes the input amino acid sequences and generates all possible subsequences starting with methionine based on the assumption that translation start sites might be mis-predicted in the original set of proteins, which in turn would result in signal peptides also being mis-predicted. The output of this step can be used as an input for secretome prediction pipelines to rescue secreted proteins with mis-predicted start sites. (Making a list and checking it twice…just like Santa does). Sequence ids for the newly generated slices are built by concatenating original sequence ids, ‘slice’ string and position of the methionine sliced from. For example: <code>ALI_PLTG_3_slice_M86</code> means that <code>ALI_PLTG_3</code> sequence was sliced starting from the methionine in the position 86.</p>
<p><code>m_slicer</code> has 2 running modes:</p>
<ul>
<li>
<code>slice</code> - to simply slice input fasta regardless of it’s origin;</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">slice &lt;-<span class="st"> </span><span class="kw"><a href="../reference/m_slicer.html">m_slicer</a></span>(aa,  <span class="co"># a set of amino acid seqeunces</span>
                <span class="dt">run_mode =</span> <span class="st">'slice'</span>,
                <span class="dt">min_len =</span> <span class="dv">100</span> <span class="co"># minimal length of the outputed slices</span>
)</code></pre></div>
<ul>
<li>
<code>rescue</code> - having output from any other up-stream function it extracts proteins not predicted to be secreted on the initial run and generate slices for them.</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">rescued &lt;-<span class="st"> </span><span class="kw"><a href="../reference/m_slicer.html">m_slicer</a></span>(step1_sp2,  <span class="co"># signalp2 output</span>
                <span class="dt">run_mode =</span> <span class="st">'rescue'</span>,
                <span class="dt">min_len =</span> <span class="dv">100</span>)</code></pre></div>
<p>Results of both run modes could be used as an input for other <code>SecretSanta</code> predictors. Say, we want to run <code>signalp</code> on the rescued proteins. Note, that <code>m_slicer</code> outputs <code>AAStringSet</code> objects, so in order to pass it to <code>signalp</code> we first need to create <code>CBSResult</code> object with <code>in_fasta = rescued</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">r_inp &lt;-<span class="st"> </span><span class="kw"><a href="../reference/SignalpResult_methods.html">SignalpResult</a></span>(<span class="dt">in_fasta =</span> rescued)
sp_rescue &lt;-<span class="st"> </span><span class="kw"><a href="../reference/signalp.html">signalp</a></span>(r_inp,
                    <span class="dt">version =</span> <span class="dv">2</span>,
                    <span class="dt">organism =</span> <span class="st">'euk'</span>,
                    <span class="dt">run_mode =</span> <span class="st">'starter'</span>,
                    <span class="dt">truncate =</span> <span class="ot">TRUE</span>)
## Version used... signalp2
## running signalp locally...
## Ok for single processing
## fasta size exceedes maximal total residue limit, seqs &gt; 416 residues will be truncated
## 206 sequences to be truncated
## Submitted sequences... 546
## signalp &lt; 4, calling parser for the output...
## Candidate sequences with signal peptides... 15</code></pre></div>
<p>The slicing procedure returned 15 additional potentially secreted candidates.</p>
</div>
</div>
<div id="pipers-andor-starters" class="section level2">
<h2 class="hasAnchor">
<a href="#pipers-andor-starters" class="anchor"></a>4. Pipers and/or starters</h2>
<p>Here is a short summary of individual methods available via the <strong>SecretSanta</strong> package:</p>
<table class="table">
<caption>
<strong>Table2</strong>: Individual methods summary.</caption>
<thead><tr class="header">
<th align="center">tool</th>
<th align="center">function</th>
<th align="left">purpouse</th>
<th align="center">organisms</th>
<th align="center">starter</th>
<th align="center">piper</th>
<th align="center">parallelised</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center">Signalp-2.0</td>
<td align="center">signalp</td>
<td align="left">predict signal peptides</td>
<td align="center">eukaryotes, gram+, gram-</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
</tr>
<tr class="even">
<td align="center">Signalp-3.0</td>
<td align="center">signalp</td>
<td align="left">predict signal peptides</td>
<td align="center">eukaryotes, gram+, gram-</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
</tr>
<tr class="odd">
<td align="center">Signalp-4.1</td>
<td align="center">signalp</td>
<td align="left">predict signal peptides</td>
<td align="center">eukaryotes, gram+, gram-</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
</tr>
<tr class="even">
<td align="center">Targetp-1.1</td>
<td align="center">targetp</td>
<td align="left">predict subcelluar localisation</td>
<td align="center">plants, not plants</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
</tr>
<tr class="odd">
<td align="center">wolfpsort</td>
<td align="center">wolfpsort</td>
<td align="left">predict subcelluar localisation</td>
<td align="center">plant, animal, fungi</td>
<td align="center">No</td>
<td align="center">Yes</td>
<td align="center">No</td>
</tr>
<tr class="even">
<td align="center">Tmhmm-2.0</td>
<td align="center">tmhmm</td>
<td align="left">predict transmembrane domains</td>
<td align="center">all taxons</td>
<td align="center">No</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
</tr>
<tr class="odd">
<td align="center">NA</td>
<td align="center">check_khdel</td>
<td align="left">scan for ER-retention motifs</td>
<td align="center">all taxons</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">No</td>
</tr>
<tr class="even">
<td align="center">NA</td>
<td align="center">m_slicer</td>
<td align="left">generate sequences with alternative translation start sites</td>
<td align="center">all taxons</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="center">No</td>
</tr>
</tbody>
</table>
</div>
<div id="build-pipelines" class="section level2">
<h2 class="hasAnchor">
<a href="#build-pipelines" class="anchor"></a>5. Build pipelines</h2>
<p>We are now ready to build a pipeline using all of our available methods. Say, we would like to have a relatively short pipeline (<strong>Figure 2</strong>) with the following analytic steps:</p>
<ul>
<li>use <strong>signalp4</strong> to predict signal peptides and cleavage sites;</li>
<li>run <strong>tmhmm</strong> on the output, to ensure that predicted proteins with signal peptides do not contain TM domains, i.e won’t be stuck in the membrane;</li>
<li>run <strong>targetp</strong> on the output, to make sure that a set of selected proteins is not targeted to plastids or mitochondria;</li>
<li>collect the targtep output and scan for ER-retention signals - to ensure that proteins won’t be stuck in the ER.</li>
</ul>
<div class="figure" style="text-align: center">
<img src="pipes_short.png" alt="\label{fig:pipe1} **Figure 3.** Minimal pipeline for secretome 
prediction" width="200px"><p class="caption">
 <strong>Figure 3.</strong> Minimal pipeline for secretome prediction
</p>
</div>
<p>Now we will run this pipeline using <code>SecretSanta</code> functions, starting with <code>aa</code> - an <code>AAStringSet</code> object containing 100 amino acid sequences.</p>
<div id="step-1-predict-signal-peptides" class="section level4">
<h4 class="hasAnchor">
<a href="#step-1-predict-signal-peptides" class="anchor"></a>Step 1: predict signal peptides</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">input &lt;-<span class="st"> </span><span class="kw"><a href="../reference/CBS_methods.html">CBSResult</a></span>(<span class="dt">in_fasta =</span> aa)

input
## An object of class "CBSResult"
## Slot "in_fasta":
##   A AAStringSet instance of length 100
##       width seq                                        names               
##   [1]    94 MVEPSSPECVVQIIKDEKLK...LLVCGFIYFAVLQSWLLES ALI_PLTG_1 Hypoth...
##   [2]   240 MTRQGYLIVHDKRSRPTVRY...VTPHYVPTHFGRGNNQEAH ALI_PLTG_2 Unknow...
##   [3]   533 MQPTSSSAPQSDVVVADPPS...QQQQEQQANSTALKNERSI ALI_PLTG_3 Hypoth...
##   [4]   456 QFLSELRKVSVSTKLDGVQR...QSQQNSRPREGCCSSCIIF ALI_PLTG_4 Hypoth...
##   [5]   210 MKLSSVCTIIFGLVFIDFNN...DGSVTTAAQKAAALVKANA ALI_PLTG_5 Cutin ...
##   ...   ... ...
##  [96]   650 MQRLVQRDKDAPVVREGDDD...DRLSLNALLEAFPPMNSRL ALI_PLTG_96 Prote...
##  [97]   236 MSTKRSRSAPILRVKKLTPE...DDDNSAKVLEWINAFIESN ALI_PLTG_97 Deoxy...
##  [98]   239 MSLNYLAHTNMRRARENFLK...KTYIEYEAVENNTLNKTEG ALI_PLTG_98 Hypot...
##  [99]   566 MCTSFSDQEDKKLVVLATEY...PRLDRKFFGLQVAPPPTSQ ALI_PLTG_99 Histo...
## [100]   180 MDNYYTSVQLLLDLQLKGLC...DPHRTFLTDLVGVELQVMG ALI_PLTG_100 Tran...
## 
## Slot "out_fasta":
##   A AAStringSet instance of length 0
step1_sp4 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/signalp.html">signalp</a></span>(input,
                    <span class="dt">version =</span> <span class="dv">4</span>,
                    <span class="dt">organism =</span> <span class="st">'euk'</span>,
                    <span class="dt">run_mode =</span> <span class="st">'starter'</span>,
                    <span class="dt">truncate =</span> <span class="ot">TRUE</span>
                    )
## Version used... signalp4
## running signalp locally...
## 2 sequences to be truncated
## Ok for single processing
## Submitted sequences... 100
## Candidate sequences with signal peptides... 5

<span class="kw"><a href="../reference/SignalpResult_methods.html">getSPtibble</a></span>(step1_sp4)
## # A tibble: 5 x 9
##       gene_id  Cmax  Cpos  Ymax  Ypos  Smax  Spos Smean     Prediction
##        &lt;fctr&gt; &lt;dbl&gt; &lt;int&gt; &lt;dbl&gt; &lt;int&gt; &lt;dbl&gt; &lt;int&gt; &lt;dbl&gt;          &lt;chr&gt;
## 1 ALI_PLTG_72 0.739    22 0.794    22 0.948     3 0.851 Signal peptide
## 2  ALI_PLTG_5 0.712    25 0.678    25 0.846    10 0.652 Signal peptide
## 3 ALI_PLTG_23 0.362    22 0.533    22 0.910    12 0.783 Signal peptide
## 4 ALI_PLTG_38 0.603    22 0.746    20 0.982    10 0.933 Signal peptide
## 5 ALI_PLTG_83 0.179    21 0.338    21 0.843    17 0.640 Signal peptide

<span class="kw"><a href="../reference/CBS_methods.html">getOutfasta</a></span>(step1_sp4)
##   A AAStringSet instance of length 5
##     width seq                                          names               
## [1]   118 MRLKFSILIFGAVLLATTTNA...NIRYASMLQDFLNTYHRRGV ALI_PLTG_72
## [2]   210 MKLSSVCTIIFGLVFIDFNNV...SDGSVTTAAQKAAALVKANA ALI_PLTG_5
## [3]   653 MKIVALVTFCIATLDSSIVFA...SEDPVYPLVKEYSDVVSKHP ALI_PLTG_23
## [4]   495 MVVRVLLVVLALLAVGVQSKA...FAIVTIGASDGRVRYMNPPS ALI_PLTG_38
## [5]   164 MVSTSRVFALLLLPSPSARIF...SMVEHIKTTKRVIDEVQDHV ALI_PLTG_83

<span class="kw"><a href="../reference/SignalpResult_methods.html">getMatfasta</a></span>(step1_sp4)
##   A AAStringSet instance of length 5
##     width seq                                          names               
## [1]    97 DDNSNKRLLRRQEKTDTTNGD...NIRYASMLQDFLNTYHRRGV ALI_PLTG_72
## [2]   186 QCSDVHVVFARGSGEAAGLGI...SDGSVTTAAQKAAALVKANA ALI_PLTG_5
## [3]   632 AECTVDELTEISTIYSEAMTD...SEDPVYPLVKEYSDVVSKHP ALI_PLTG_23
## [4]   474 TRVRKSWTAYTSDEKEIYLSA...FAIVTIGASDGRVRYMNPPS ALI_PLTG_38
## [5]   144 FQQVTMSATTATAITTKLKQF...SMVEHIKTTKRVIDEVQDHV ALI_PLTG_83</code></pre></div>
<hr>
<blockquote>
<p><strong>Result</strong>: 5 candidate proteins with signal peptides.</p>
</blockquote>
</div>
<div id="step-2-check-for-presence-of-tm-domains-in-mature-peptides" class="section level4">
<h4 class="hasAnchor">
<a href="#step-2-check-for-presence-of-tm-domains-in-mature-peptides" class="anchor"></a>Step 2: check for presence of TM domains in mature peptides:</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># we allow 0 TM domains in mature peptides</span>
step2_tm &lt;-<span class="st"> </span><span class="kw"><a href="../reference/tmhmm.html">tmhmm</a></span>(step1_sp4, <span class="dt">TM =</span> <span class="dv">0</span>)
## running TMHMM locally...
## Submitted sequences... 5
## Candidates with signal peptides and 0 TM domains in mature seq 5

<span class="kw"><a href="../reference/TMhmmResult_methods.html">getTMtibble</a></span>(step2_tm)
## # A tibble: 5 x 6
##       gene_id length ExpAA First60 PredHel Topology
##         &lt;chr&gt;  &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;    &lt;chr&gt;
## 1 ALI_PLTG_72     97  0.00    0.00       0        i
## 2  ALI_PLTG_5    186  1.29    1.16       0        o
## 3 ALI_PLTG_23    632  0.00    0.00       0        o
## 4 ALI_PLTG_38    474  0.09    0.06       0        o
## 5 ALI_PLTG_83    144  1.47    0.00       0        o

<span class="kw"><a href="../reference/CBS_methods.html">getOutfasta</a></span>(step2_tm)
##   A AAStringSet instance of length 5
##     width seq                                          names               
## [1]   118 MRLKFSILIFGAVLLATTTNA...NIRYASMLQDFLNTYHRRGV ALI_PLTG_72
## [2]   210 MKLSSVCTIIFGLVFIDFNNV...SDGSVTTAAQKAAALVKANA ALI_PLTG_5
## [3]   653 MKIVALVTFCIATLDSSIVFA...SEDPVYPLVKEYSDVVSKHP ALI_PLTG_23
## [4]   495 MVVRVLLVVLALLAVGVQSKA...FAIVTIGASDGRVRYMNPPS ALI_PLTG_38
## [5]   164 MVSTSRVFALLLLPSPSARIF...SMVEHIKTTKRVIDEVQDHV ALI_PLTG_83

<span class="kw"><a href="../reference/TMhmmResult_methods.html">getOutMatfasta</a></span>(step2_tm)
##   A AAStringSet instance of length 5
##     width seq                                          names               
## [1]    97 DDNSNKRLLRRQEKTDTTNGD...NIRYASMLQDFLNTYHRRGV ALI_PLTG_72
## [2]   186 QCSDVHVVFARGSGEAAGLGI...SDGSVTTAAQKAAALVKANA ALI_PLTG_5
## [3]   632 AECTVDELTEISTIYSEAMTD...SEDPVYPLVKEYSDVVSKHP ALI_PLTG_23
## [4]   474 TRVRKSWTAYTSDEKEIYLSA...FAIVTIGASDGRVRYMNPPS ALI_PLTG_38
## [5]   144 FQQVTMSATTATAITTKLKQF...SMVEHIKTTKRVIDEVQDHV ALI_PLTG_83</code></pre></div>
<hr>
<blockquote>
<p><strong>Result</strong>: 5 candidate proteins with signal peptides; 0 TM domains in mature sequences.</p>
</blockquote>
</div>
<div id="step-3-predict-sub-cellular-localization" class="section level4">
<h4 class="hasAnchor">
<a href="#step-3-predict-sub-cellular-localization" class="anchor"></a>Step 3: predict sub-cellular localization</h4>
<p>Here we are using <code>targetp</code>, we could have also used <code>wolfpsort</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">step3_tp &lt;-<span class="st"> </span><span class="kw"><a href="../reference/targetp.html">targetp</a></span>(step2_tm, <span class="dt">network =</span> <span class="st">'N'</span>,<span class="dt">run_mode =</span> <span class="st">'piper'</span>)
## running targetp locally...
## Ok for single processing
## Number of submitted sequences... 5
## Number of candidate secreted sequences 4

<span class="kw"><a href="../reference/TargetpResult_methods.html">getTPtibble</a></span>(step3_tp)
## # A tibble: 4 x 7
##       gene_id length   mTP    sp other TP_localization    RC
##        &lt;fctr&gt;  &lt;int&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;          &lt;fctr&gt; &lt;int&gt;
## 1 ALI_PLTG_72    118 0.086 0.877 0.054               S     2
## 2  ALI_PLTG_5    210 0.027 0.944 0.096               S     1
## 3 ALI_PLTG_23    653 0.025 0.953 0.072               S     1
## 4 ALI_PLTG_38    495 0.123 0.951 0.012               S     1

<span class="kw"><a href="../reference/CBS_methods.html">getInfasta</a></span>(step3_tp)
##   A AAStringSet instance of length 5
##     width seq                                          names               
## [1]   118 MRLKFSILIFGAVLLATTTNA...NIRYASMLQDFLNTYHRRGV ALI_PLTG_72
## [2]   210 MKLSSVCTIIFGLVFIDFNNV...SDGSVTTAAQKAAALVKANA ALI_PLTG_5
## [3]   653 MKIVALVTFCIATLDSSIVFA...SEDPVYPLVKEYSDVVSKHP ALI_PLTG_23
## [4]   495 MVVRVLLVVLALLAVGVQSKA...FAIVTIGASDGRVRYMNPPS ALI_PLTG_38
## [5]   164 MVSTSRVFALLLLPSPSARIF...SMVEHIKTTKRVIDEVQDHV ALI_PLTG_83

<span class="kw"><a href="../reference/CBS_methods.html">getOutfasta</a></span>(step3_tp)
##   A AAStringSet instance of length 4
##     width seq                                          names               
## [1]   118 MRLKFSILIFGAVLLATTTNA...NIRYASMLQDFLNTYHRRGV ALI_PLTG_72
## [2]   210 MKLSSVCTIIFGLVFIDFNNV...SDGSVTTAAQKAAALVKANA ALI_PLTG_5
## [3]   653 MKIVALVTFCIATLDSSIVFA...SEDPVYPLVKEYSDVVSKHP ALI_PLTG_23
## [4]   495 MVVRVLLVVLALLAVGVQSKA...FAIVTIGASDGRVRYMNPPS ALI_PLTG_38</code></pre></div>
<p>This step allowed us to filter 1 sequence potentially targeted to mitochondria.</p>
<blockquote>
<p><strong>Result</strong>: 4 candidate proteins with signal peptides; 0 TM domains in mature sequences; not targeted to plastids or mitochondria.</p>
</blockquote>
</div>
<div id="step-4-check-for-er-retention-signals" class="section level4">
<h4 class="hasAnchor">
<a href="#step-4-check-for-er-retention-signals" class="anchor"></a>Step 4: check for ER-retention signals</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">step4_er &lt;-<span class="st"> </span><span class="kw"><a href="../reference/check_khdel.html">check_khdel</a></span>(step3_tp,
                        <span class="dt">run_mode =</span> <span class="st">'piper'</span>)
## checking for terminal ER retention signals...
## Submitted sequences... 4
## Sequences with terminal ER retention signals detected... 0
## Candidate without terminal ER retention signals detected... 4

<span class="kw"><a href="../reference/CBS_methods.html">getOutfasta</a></span>(step4_er)
##   A AAStringSet instance of length 4
##     width seq                                          names               
## [1]   118 MRLKFSILIFGAVLLATTTNA...NIRYASMLQDFLNTYHRRGV ALI_PLTG_72
## [2]   210 MKLSSVCTIIFGLVFIDFNNV...SDGSVTTAAQKAAALVKANA ALI_PLTG_5
## [3]   653 MKIVALVTFCIATLDSSIVFA...SEDPVYPLVKEYSDVVSKHP ALI_PLTG_23
## [4]   495 MVVRVLLVVLALLAVGVQSKA...FAIVTIGASDGRVRYMNPPS ALI_PLTG_38</code></pre></div>
<hr>
<blockquote>
<p><strong>Final Result</strong>: 4 candidate proteins with signal peptides; 0 TM domains in mature sequences; not targeted to plastids or mitochondria; not retained in ER.</p>
</blockquote>
<p>This is an example of a fairly simple pipeline, but you can create more complex ones. A more stringent pipeline might involve multi-step filtration on signal peptide prediction as well as additional predictions for ‘rescued’ (<strong>Figure 4</strong>) sequences.</p>
<div class="figure" style="text-align: center">
<img src="pipes_long.png" alt="\label{fig:pipe1} **Figure 4.** Example of a stringent pipeline." width="300px"><p class="caption">
 <strong>Figure 4.</strong> Example of a stringent pipeline.
</p>
</div>
</div>
<div id="step-5-save-the-result-secretome" class="section level4">
<h4 class="hasAnchor">
<a href="#step-5-save-the-result-secretome" class="anchor"></a>Step 5: save the result secretome</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">writeXStringSet</span>(<span class="kw"><a href="../reference/CBS_methods.html">getOutfasta</a></span>(step4_er), <span class="st">'secretome_out.fasta'</span>)</code></pre></div>
</div>
<div id="compact-pipelienes" class="section level4">
<h4 class="hasAnchor">
<a href="#compact-pipelienes" class="anchor"></a>Compact pipelienes:</h4>
<p>Now, once we have examined all the individual analytic steps, we can directly combine them in a single command by using <code>%&gt;%</code> operator:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pipeline &lt;-<span class="st"> </span><span class="kw"><a href="../reference/signalp.html">signalp</a></span>(input, <span class="dt">version =</span> <span class="dv">4</span>, <span class="dt">organism =</span> <span class="st">'euk'</span>, <span class="dt">run_mode =</span> <span class="st">'starter'</span>) %&gt;%<span class="st"> </span>
<span class="st">    </span><span class="kw"><a href="../reference/tmhmm.html">tmhmm</a></span>(<span class="dt">TM =</span> <span class="dv">0</span>) %&gt;%
<span class="st">    </span><span class="kw"><a href="../reference/targetp.html">targetp</a></span>(<span class="dt">network =</span> <span class="st">'N'</span>, <span class="dt">run_mode =</span> <span class="st">'piper'</span>) %&gt;%
<span class="st">    </span><span class="kw"><a href="../reference/check_khdel.html">check_khdel</a></span>(<span class="dt">run_mode =</span> <span class="st">'piper'</span>)
## Version used... signalp4
## running signalp locally...
## 2 sequences to be truncated
## Ok for single processing
## Submitted sequences... 100
## Candidate sequences with signal peptides... 5
## running TMHMM locally...
## Submitted sequences... 5
## Candidates with signal peptides and 0 TM domains in mature seq 5
## running targetp locally...
## Ok for single processing
## Number of submitted sequences... 5
## Number of candidate secreted sequences 4
## checking for terminal ER retention signals...
## Submitted sequences... 4
## Sequences with terminal ER retention signals detected... 0
## Candidate without terminal ER retention signals detected... 4</code></pre></div>
<p>Since <code>pp</code> belongs to the <code>CBSResult</code> class, we can again extarct the output fasta field and save it in a separate file:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">writeXStringSet</span>(<span class="kw"><a href="../reference/CBS_methods.html">getOutfasta</a></span>(pipeline), <span class="st">'secretome_out.fasta'</span>)</code></pre></div>
</div>
<div id="session" class="section level4">
<h4 class="hasAnchor">
<a href="#session" class="anchor"></a>Session</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sessionInfo</span>()</code></pre></div>
<pre><code>## R version 3.4.0 (2017-04-21)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 16.04.2 LTS
## 
## Matrix products: default
## BLAS: /usr/lib/libblas/libblas.so.3.6.0
## LAPACK: /usr/lib/lapack/liblapack.so.3.6.0
## 
## locale:
##  [1] LC_CTYPE=en_GB.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_GB.UTF-8        LC_COLLATE=en_GB.UTF-8    
##  [5] LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_GB.UTF-8   
##  [7] LC_PAPER=en_GB.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
## [1] bindrcpp_0.2        SecretSanta_0.99.0  dplyr_0.7.4        
## [4] Biostrings_2.44.2   XVector_0.16.0      IRanges_2.10.3     
## [7] S4Vectors_0.14.4    BiocGenerics_0.22.0
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_0.12.13     knitr_1.17       bindr_0.1        magrittr_1.5    
##  [5] hms_0.3          zlibbioc_1.22.0  R6_2.2.2         rlang_0.1.2.9000
##  [9] stringr_1.2.0    highr_0.6        tools_3.4.0      htmltools_0.3.6 
## [13] yaml_2.1.14      rprojroot_1.2    digest_0.6.12    assertthat_0.2.0
## [17] tibble_1.3.4     readr_1.1.1      glue_1.1.1       evaluate_0.10.1 
## [21] rmarkdown_1.6    stringi_1.1.5    compiler_3.4.0   backports_1.1.0 
## [25] pkgconfig_2.0.1</code></pre>
</div>
<div id="references" class="section level3 unnumbered">
<h3 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h3>
<div id="refs" class="references">
<div id="ref-Bendtsen2004">
<p>Dyrl�v Bendtsen, Gunnar von Heijne annick, Henrik Nielsen, and S�ren Brunak. 2004. “Improved Prediction of Signal Peptides: SignalP 3.0.” <em>J. Mol. Biol.</em>, no. 340: 783–95.</p>
</div>
<div id="ref-Evangelisti2017">
<p>Evangelisti E, Hainaux T, Gogleva A. 2017. “Time-Resolved Dual Transcriptomics Reveal Early Induced Nicotiana Benthamiana Root Genes and Conserved Infection-Promoting Phytophthora Palmivora Effectors.” <em>BMC Biology</em>, no. 15:39.</p>
</div>
<div id="ref-Nielsen1997">
<p>Henrik Nielsen, S�ren Brunak, Jacob Engelbrecht, and Gunnar von Heijne. 1997. “Identification of Prokaryotic and Eukaryotic Signal Peptides and Prediction of Their Cleavage Sites.” <em>Protein Engineering</em> 10: 1–6.</p>
</div>
<div id="ref-Krogh2001">
<p>Krogh, Anders, Bjorn Larsson Gunnar von Heijne, and Erik L.L. Sonnhammer. 2001. “Predicting Transmembrane Protein Topology with a Hidden Markov Model: Application to Complete Genomes.” <em>J. Mol. Biol.</em>, no. 305: 567–80.</p>
</div>
<div id="ref-Munro1987">
<p>Munro S, Pelham HBR. 1987. “A c-Terminal Signal Prevents Secretion of Luminal Er Proteins.” <em>Cell</em>, no. 48: 899–907.</p>
</div>
<div id="ref-Nielsen1998">
<p>Nielsen, Henrik, and Anders Krogh. 1998. “Prediction of Signal Peptides and Signal Anchors by a Hidden Markov Model.” <em>Proceedings of the Sixth International Conference on Intelligent Systems for Molecular Biology (ISMB 6)</em>, 122–30.</p>
</div>
<div id="ref-Emanuelsson2008">
<p>Olof Emanuelsson, Soren Brunak, Henrik Nielsen, and Gunnar von Heijne. 2008. “Predicting Subcellular Localization of Proteins Based on Their N-Terminal Amino Acid Sequence.” <em>J. Mol. Biol.</em>, no. 300: 1005–16.</p>
</div>
<div id="ref-Horton2006">
<p>Paul Horton, Takeshi Obayashi, Keun-Joon Park, and Kenta Nakai. 2006. “Protein Subcellular Localization Prediction with Wolf Psort.” <em>Asian Pacific Bioinformatics Conference, APCB2006</em>.</p>
</div>
<div id="ref-Sperschneider2015">
<p>Sperschneider J, Hane JK, Williams AH. 2015. “Evaluation of Secretion Prediction Highlights Differing Approaches Needed for Oomycete and Fungal Effectors.” <em>Front Plant Sci</em>, no. 6: 1168.</p>
</div>
<div id="ref-Petersen2011">
<p>Thomas Nordahl Petersen, Gunnar von Heijne, Soren Brunak, and Henrik Nielsen. 2011. “SignalP 4.0: Discriminating Signal Peptides from Transmembrane Regions.” <em>Nature Methods</em> 8 (10): 785–86.</p>
</div>
</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#background">1. Background</a></li>
      <li><a href="#installation-of-external-dependencies">2. Installation of external dependencies</a></li>
      <li><a href="#individual-methods">3. Individual methods</a></li>
      <li><a href="#pipers-andor-starters">4. Pipers and/or starters</a></li>
      <li><a href="#build-pipelines">5. Build pipelines</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Anna Gogleva.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
